From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: qmuntal <qmuntaldiaz@microsoft.com>
Date: Wed, 15 Jan 2025 16:32:26 +0100
Subject: [PATCH] Implement crypto/internal/backend

Implement the supporting code for using the crypto backends,
but without using them yet.

Update the Go toolchain build settings so that it uses the
desired goexperiments and build tags.
---
 .gitignore                                    |   2 +
 .../compile/internal/logopt/logopt_test.go    |   6 +-
 .../compile/internal/ssa/stmtlines_test.go    |   1 -
 src/cmd/compile/script_test.go                |   8 +
 src/cmd/dist/build.go                         |  11 +
 src/cmd/dist/test.go                          |  44 ++-
 src/cmd/go/go_test.go                         |  11 +
 src/cmd/go/internal/load/pkg.go               |   3 +
 src/cmd/go/systemcrypto_test.go               | 144 +++++++
 .../go/testdata/script/env_cross_build.txt    |   2 +
 .../script/test_android_issue62123.txt        |   2 +
 src/cmd/internal/testdir/testdir_test.go      |   7 +
 src/cmd/link/link_test.go                     |   8 +
 src/crypto/internal/backend/backend_test.go   |  30 ++
 src/crypto/internal/backend/bbig/big.go       |  17 +
 src/crypto/internal/backend/bbig/big_cng.go   |  12 +
 .../internal/backend/bbig/big_darwin.go       |  12 +
 .../internal/backend/bbig/big_openssl.go      |  12 +
 src/crypto/internal/backend/cng_windows.go    | 345 ++++++++++++++++
 src/crypto/internal/backend/common.go         |  46 +++
 src/crypto/internal/backend/darwin_darwin.go  | 372 ++++++++++++++++++
 src/crypto/internal/backend/fips140/cng.go    |  35 ++
 src/crypto/internal/backend/fips140/darwin.go |  13 +
 .../internal/backend/fips140/fips140.go       |  70 ++++
 .../internal/backend/fips140/isrequirefips.go |   9 +
 .../internal/backend/fips140/norequirefips.go |   9 +
 .../backend/fips140/nosystemcrypto.go         |  13 +
 .../internal/backend/fips140/openssl_cgo.go   |  59 +++
 .../internal/backend/fips140/openssl_nocgo.go |  17 +
 .../fips140/requirefips_nosystemcrypto.go     |  15 +
 .../backend/fips140/skipfipscheck_off.go      |   9 +
 .../backend/fips140/skipfipscheck_on.go       |   9 +
 .../internal/opensslsetup/opensslsetup.go     |  70 ++++
 .../opensslsetup/opensslsetup_test.go         |  92 +++++
 .../backend/internal/opensslsetup/stub.go     |   8 +
 src/crypto/internal/backend/nobackend.go      | 251 ++++++++++++
 src/crypto/internal/backend/openssl_linux.go  | 332 ++++++++++++++++
 src/crypto/internal/backend/stub.s            |  10 +
 src/crypto/systemcrypto_nocgo.go              |  15 +
 src/go/build/deps_test.go                     |  32 +-
 src/internal/buildcfg/exp.go                  |  54 +++
 src/runtime/runtime_boring.go                 |   5 +
 src/syscall/exec_linux_test.go                |   2 +-
 43 files changed, 2207 insertions(+), 17 deletions(-)
 create mode 100644 src/cmd/go/systemcrypto_test.go
 create mode 100644 src/crypto/internal/backend/backend_test.go
 create mode 100644 src/crypto/internal/backend/bbig/big.go
 create mode 100644 src/crypto/internal/backend/bbig/big_cng.go
 create mode 100644 src/crypto/internal/backend/bbig/big_darwin.go
 create mode 100644 src/crypto/internal/backend/bbig/big_openssl.go
 create mode 100644 src/crypto/internal/backend/cng_windows.go
 create mode 100644 src/crypto/internal/backend/common.go
 create mode 100644 src/crypto/internal/backend/darwin_darwin.go
 create mode 100644 src/crypto/internal/backend/fips140/cng.go
 create mode 100644 src/crypto/internal/backend/fips140/darwin.go
 create mode 100644 src/crypto/internal/backend/fips140/fips140.go
 create mode 100644 src/crypto/internal/backend/fips140/isrequirefips.go
 create mode 100644 src/crypto/internal/backend/fips140/norequirefips.go
 create mode 100644 src/crypto/internal/backend/fips140/nosystemcrypto.go
 create mode 100644 src/crypto/internal/backend/fips140/openssl_cgo.go
 create mode 100644 src/crypto/internal/backend/fips140/openssl_nocgo.go
 create mode 100644 src/crypto/internal/backend/fips140/requirefips_nosystemcrypto.go
 create mode 100644 src/crypto/internal/backend/fips140/skipfipscheck_off.go
 create mode 100644 src/crypto/internal/backend/fips140/skipfipscheck_on.go
 create mode 100644 src/crypto/internal/backend/internal/opensslsetup/opensslsetup.go
 create mode 100644 src/crypto/internal/backend/internal/opensslsetup/opensslsetup_test.go
 create mode 100644 src/crypto/internal/backend/internal/opensslsetup/stub.go
 create mode 100644 src/crypto/internal/backend/nobackend.go
 create mode 100644 src/crypto/internal/backend/openssl_linux.go
 create mode 100644 src/crypto/internal/backend/stub.s
 create mode 100644 src/crypto/systemcrypto_nocgo.go

diff --git a/.gitignore b/.gitignore
index c6512e64a4ef39..b3b01db73b009d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -46,6 +46,8 @@ _testmain.go
 /test/run.out
 /test/times.out
 
+!/src/vendor/github.com/microsoft/go-crypto-darwin/internal/cryptokit/CryptoKit.o
+
 # This file includes artifacts of Go build that should not be checked in.
 # For files created by specific development environment (e.g. editor),
 # use alternative ways to exclude files from git.
diff --git a/src/cmd/compile/internal/logopt/logopt_test.go b/src/cmd/compile/internal/logopt/logopt_test.go
index 1edabf9fb7ff04..026f8cb0278ab6 100644
--- a/src/cmd/compile/internal/logopt/logopt_test.go
+++ b/src/cmd/compile/internal/logopt/logopt_test.go
@@ -5,6 +5,7 @@
 package logopt
 
 import (
+	"internal/goexperiment"
 	"internal/testenv"
 	"os"
 	"path/filepath"
@@ -147,7 +148,10 @@ func s15a8(x *[15]int64) [15]int64 {
 		arches := []string{runtime.GOARCH}
 		goos0 := runtime.GOOS
 		if runtime.GOARCH == "amd64" { // Test many things with "linux" (wasm will get "js")
-			arches = []string{"arm", "arm64", "386", "amd64", "mips", "mips64", "loong64", "ppc64le", "riscv64", "s390x", "wasm"}
+			if !goexperiment.SystemCrypto {
+				// SystemCrypto does not support cross-compilation.
+				arches = []string{"arm", "arm64", "386", "amd64", "mips", "mips64", "loong64", "ppc64le", "riscv64", "s390x", "wasm"}
+			}
 			goos0 = "linux"
 		}
 
diff --git a/src/cmd/compile/internal/ssa/stmtlines_test.go b/src/cmd/compile/internal/ssa/stmtlines_test.go
index e17a5402af818d..e83c4c5b79d11b 100644
--- a/src/cmd/compile/internal/ssa/stmtlines_test.go
+++ b/src/cmd/compile/internal/ssa/stmtlines_test.go
@@ -113,7 +113,6 @@ func TestStmtLines(t *testing.T) {
 		must(err)
 
 		var le dwarf.LineEntry
-
 		for {
 			err := lrdr.Next(&le)
 			if err == io.EOF {
diff --git a/src/cmd/compile/script_test.go b/src/cmd/compile/script_test.go
index 0e32e0769ee2e2..eda2d576acc6c6 100644
--- a/src/cmd/compile/script_test.go
+++ b/src/cmd/compile/script_test.go
@@ -7,6 +7,7 @@ package main
 import (
 	"cmd/internal/script/scripttest"
 	"flag"
+	"internal/goexperiment"
 	"internal/testenv"
 	"os"
 	"runtime"
@@ -50,6 +51,13 @@ func TestScript(t *testing.T) {
 		// the installed linker with our test binary.
 		doReplacement = false
 	}
+	if goexperiment.SystemCrypto {
+		// If systemcrypto is enabled, we can't replace the compiler
+		// because the test binary is built with systemcrypto, but the
+		// compiler is not. This would cause a mismatch in the
+		// compiler version check in the script tests.
+		doReplacement = false
+	}
 	repls := []scripttest.ToolReplacement{}
 	if doReplacement {
 		if testCompiler == "" {
diff --git a/src/cmd/dist/build.go b/src/cmd/dist/build.go
index 024050c2dd70c4..3c82b7436d10bf 100644
--- a/src/cmd/dist/build.go
+++ b/src/cmd/dist/build.go
@@ -1387,6 +1387,12 @@ func toolenv() []string {
 		// still have full paths for stack traces for compiler crashes and the like.
 		env = append(env, "GOFLAGS=-trimpath -ldflags=-w -gcflags=cmd/...=-dwarf=false")
 	}
+	if goexp := os.Getenv("GOEXPERIMENT"); goexp != "" {
+		// Remove systemcrypto from GOEXPERIMENT: to favor portability, the
+		// Microsoft build of Go should not use it. Go only uses crypto packages
+		// for non-crypto purposes, such as generating build IDs.
+		env = append(env, "GOEXPERIMENT="+goexp+",nosystemcrypto")
+	}
 	return env
 }
 
@@ -1548,6 +1554,11 @@ func cmdbootstrap() {
 	os.Setenv("CC", compilerEnvLookup("CC", defaultcc, goos, goarch))
 	// Now that cmd/go is in charge of the build process, enable GOEXPERIMENT.
 	os.Setenv("GOEXPERIMENT", goexperiment)
+	// Build the Go toolchain with "-tags=ms_skipfipscheck". This
+	// allows toolchains not built with the systemcrypto goexperiment to be used
+	// when GODEBUG=fips140=on is set. For example, when running
+	// "GODEBUG=fips140=on go test ./..." or "GODEBUG=fips140=on go run .".
+	os.Setenv("GOFLAGS", "-tags=ms_skipfipscheck")
 	// No need to enable PGO for toolchain2.
 	goInstall(toolenv(), goBootstrap, append([]string{"-pgo=off"}, toolchain...)...)
 	if debug {
diff --git a/src/cmd/dist/test.go b/src/cmd/dist/test.go
index aa09d1eba34be8..c3f3a8324c507c 100644
--- a/src/cmd/dist/test.go
+++ b/src/cmd/dist/test.go
@@ -156,10 +156,12 @@ func (t *tester) run() {
 		}
 	}
 
+	tags := []string{"-tags=ms_skipfipscheck"}
+
 	if t.rebuild {
 		t.out("Building packages and commands.")
 		// Force rebuild the whole toolchain.
-		goInstall(toolenv(), gorootBinGo, append([]string{"-a"}, toolchain...)...)
+		goInstall(toolenv(), gorootBinGo, append([]string{"-a"}, append(tags, toolchain...)...)...)
 	}
 
 	if !t.listMode {
@@ -176,9 +178,9 @@ func (t *tester) run() {
 			// and virtualization we usually start with a clean GOCACHE, so we would
 			// end up rebuilding large parts of the standard library that aren't
 			// otherwise relevant to the actual set of packages under test.
-			goInstall(toolenv(), gorootBinGo, toolchain...)
-			goInstall(toolenv(), gorootBinGo, toolchain...)
-			goInstall(toolenv(), gorootBinGo, toolsToInstall...)
+			goInstall(toolenv(), gorootBinGo, append(tags, toolchain...)...)
+			goInstall(toolenv(), gorootBinGo, append(tags, toolchain...)...)
+			goInstall(toolenv(), gorootBinGo, append(tags, "cmd")...)
 		}
 	}
 
@@ -749,11 +751,18 @@ func (t *tester) registerTests() {
 			variant: "jsonv2",
 			env:     []string{"GOEXPERIMENT=jsonv2"},
 			pkg:     "encoding/json/...",
+			// Skip FIPS checks. The test overrides GOEXPERIMENT,
+			// so if we have been executed with GOEXPERIMENT=systemcrypto
+			// and FIPS mode is required, then the test will fail because
+			// the systemcrypto package will not be available.
+			tags: []string{"ms_skipfipscheck"},
 		})
 	}
 
 	// Test ios/amd64 for the iOS simulator.
-	if goos == "darwin" && goarch == "amd64" && t.cgoEnabled {
+	if goos == "darwin" && goarch == "amd64" && t.cgoEnabled &&
+		// IOS is not supported with darwincrypto.
+		!(strings.Contains(goexperiment, "systemcrypto") || strings.Contains(goexperiment, "darwincrypto")) {
 		t.registerTest("GOOS=ios on darwin/amd64",
 			&goTest{
 				variant:  "amd64ios",
@@ -863,14 +872,21 @@ func (t *tester) registerTests() {
 
 	// Test internal linking of PIE binaries where it is supported.
 	if t.internalLinkPIE() && !disablePIE {
+		var tags, env []string
+		if goos != "windows" {
+			// Linux and macOS don't support systemcrypto with CGO_ENABLED=0.
+			tags = append(tags, "ms_skipfipscheck")
+			env = append(env, "GOEXPERIMENT=nosystemcrypto")
+		}
 		t.registerTest("internal linking, -buildmode=pie",
 			&goTest{
 				variant:   "pie_internal",
 				timeout:   60 * time.Second,
 				buildmode: "pie",
 				ldflags:   "-linkmode=internal",
-				env:       []string{"CGO_ENABLED=0"},
+				env:       append([]string{"CGO_ENABLED=0"}, env...),
 				pkg:       "reflect",
+				tags:      tags,
 			})
 		t.registerTest("internal linking, -buildmode=pie",
 			&goTest{
@@ -878,9 +894,10 @@ func (t *tester) registerTests() {
 				timeout:   60 * time.Second,
 				buildmode: "pie",
 				ldflags:   "-linkmode=internal",
-				env:       []string{"CGO_ENABLED=0"},
+				env:       append([]string{"CGO_ENABLED=0"}, env...),
 				pkg:       "crypto/internal/fips140test",
 				runTests:  "TestFIPSCheck",
+				tags:      tags,
 			})
 		// Also test a cgo package.
 		if t.cgoEnabled && t.internalLink() && !disablePIE {
@@ -897,15 +914,22 @@ func (t *tester) registerTests() {
 
 	if t.extLink() && !t.compileOnly {
 		if goos != "android" { // Android does not support non-PIE linking
+			var tags, env []string
+			if goos == "darwin" {
+				// Darwincrypto doesn't support external linking.
+				tags = append(tags, "ms_skipfipscheck")
+				env = append(env, "GOEXPERIMENT=nosystemcrypto")
+			}
 			t.registerTest("external linking, -buildmode=exe",
 				&goTest{
 					variant:   "exe_external",
 					timeout:   60 * time.Second,
 					buildmode: "exe",
 					ldflags:   "-linkmode=external",
-					env:       []string{"CGO_ENABLED=1"},
+					env:       append([]string{"CGO_ENABLED=1"}, env...),
 					pkg:       "crypto/internal/fips140test",
 					runTests:  "TestFIPSCheck",
+					tags:      tags,
 				})
 		}
 		if t.externalLinkPIE() && !disablePIE {
@@ -984,7 +1008,9 @@ func (t *tester) registerTests() {
 		t.registerRaceTests()
 	}
 
-	if goos != "android" && !t.iOS() {
+	// cmd/internal/testdir uses -buildmode=exe on darwin,
+	// which is not supported by darwincrypto.
+	if goos != "android" && !t.iOS() && !(strings.Contains(goexperiment, "systemcrypto") || strings.Contains(goexperiment, "darwincrypto")) {
 		// Only start multiple test dir shards on builders,
 		// where they get distributed to multiple machines.
 		// See issues 20141 and 31834.
diff --git a/src/cmd/go/go_test.go b/src/cmd/go/go_test.go
index 3e691abe41f702..b799031d598892 100644
--- a/src/cmd/go/go_test.go
+++ b/src/cmd/go/go_test.go
@@ -14,6 +14,7 @@ import (
 	"fmt"
 	"go/format"
 	"internal/godebug"
+	"internal/goexperiment"
 	"internal/platform"
 	"internal/testenv"
 	"io"
@@ -111,6 +112,13 @@ func TestMain(m *testing.M) {
 		if v := os.Getenv("TESTGO_TOOLCHAIN_VERSION"); v != "" {
 			work.ToolchainVersion = v
 		}
+		// Remove the " X:systemcrypto" suffix from the toolchain version,
+		// else it won't match the compiler version and the tests will fail.
+		// This happens because "go" invocations inside TestScript don't
+		// really call the go binary, which is never built with systemcrypto,
+		// but it instead calls the test binary with CMDGO_TEST_RUN_MAIN=1,
+		// which can be built with systemcrypto.
+		work.ToolchainVersion = strings.ReplaceAll(work.ToolchainVersion, " X:systemcrypto", "")
 
 		if testGOROOT := os.Getenv("TESTGO_GOROOT"); testGOROOT != "" {
 			// Disallow installs to the GOROOT from which testgo was built.
@@ -1861,6 +1869,9 @@ func TestGenerateUsesBuildContext(t *testing.T) {
 }
 
 func TestGoEnv(t *testing.T) {
+	if goexperiment.SystemCrypto {
+		t.Skip("freebsd is not supported by system crypto")
+	}
 	tg := testgo(t)
 	tg.parallel()
 	defer tg.cleanup()
diff --git a/src/cmd/go/internal/load/pkg.go b/src/cmd/go/internal/load/pkg.go
index e913f9885234fd..ef331980a816a3 100644
--- a/src/cmd/go/internal/load/pkg.go
+++ b/src/cmd/go/internal/load/pkg.go
@@ -2412,6 +2412,9 @@ func (p *Package) setBuildInfo(ctx context.Context, autoVCS bool) {
 			buildmode = "archive"
 		}
 	}
+	if cfg.Experiment.SystemCrypto {
+		appendSetting("microsoft_systemcrypto", "1")
+	}
 	appendSetting("-buildmode", buildmode)
 	appendSetting("-compiler", cfg.BuildContext.Compiler)
 	if gccgoflags := BuildGccgoflags.String(); gccgoflags != "" && cfg.BuildContext.Compiler == "gccgo" {
diff --git a/src/cmd/go/systemcrypto_test.go b/src/cmd/go/systemcrypto_test.go
new file mode 100644
index 00000000000000..33d85d0df477d7
--- /dev/null
+++ b/src/cmd/go/systemcrypto_test.go
@@ -0,0 +1,144 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main_test
+
+import (
+	"internal/testenv"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+)
+
+const fileWithCrypto = `
+package main
+
+import (
+	"crypto/sha256"
+	"fmt"
+)
+
+func main() {
+	fmt.Println(sha256.Sum256([]byte("Hello, World!")))
+}
+`
+
+const fileWithoutCrypto = `
+package main
+
+import (
+	"fmt"
+)
+
+func main() {
+	fmt.Println("Hello, World!")
+}
+`
+
+func execGoTool(t *testing.T, allowFail bool, env []string, args ...string) (bool, string) {
+	t.Helper()
+	cmd := testenv.Command(t, testenv.GoToolPath(t), args...)
+	cmd = testenv.CleanCmdEnv(cmd)
+	cmd.Env = append(cmd.Env, env...)
+	out, err := cmd.CombinedOutput()
+	sout := strings.TrimSpace(string(out))
+	if err != nil {
+		if allowFail {
+			return false, sout
+		}
+		t.Fatalf("go build failed: %v\n%s", err, out)
+	}
+	return true, sout
+}
+
+// writeFile creates a temporary file with the given content and returns its path.
+func writeFile(t *testing.T, content string) string {
+	t.Helper()
+	name := "main.go"
+	if strings.Contains(content, "testing") {
+		name = "main_test.go"
+	}
+	name = filepath.Join(t.TempDir(), name)
+	if err := os.WriteFile(name, []byte(content), 0o644); err != nil {
+		t.Fatal(err)
+	}
+	return name
+}
+
+func outPath(t *testing.T) string {
+	t.Helper()
+	return filepath.Join(t.TempDir(), "out")
+}
+
+func TestSystemCryptoFIPS(t *testing.T) {
+	// Test different go commands with GODEBUG=fips140=on
+	// to exercise the ms_skipfipscheck build tag.
+	t.Parallel()
+	env := []string{"GODEBUG=fips140=on"}
+
+	cryptoFile := writeFile(t, fileWithCrypto)
+	nonCryptoFile := writeFile(t, fileWithoutCrypto)
+
+	// Build should always succeed given that the go toolchain
+	// is built with the ms_skipfipscheck build tag.
+	execGoTool(t, false, env, "build", "-o", outPath(t), cryptoFile)
+	execGoTool(t, false, env, "build", "-o", outPath(t), nonCryptoFile)
+
+	// Run should always succeed if the target go package
+	// doesn't use crypto.
+	execGoTool(t, false, env, "run", nonCryptoFile)
+
+	// Run may or may not fail if the target go package uses crypto,
+	// because while the toolchain was built with ms_skipfipscheck, the
+	// target program was not. Failure depends on system crypto mode
+	// and presence of system-provided crypto, and it can't be tested here.
+}
+
+func TestSystemCryptoSetting(t *testing.T) {
+	t.Parallel()
+	file := writeFile(t, fileWithCrypto)
+
+	type testCase struct {
+		enabled bool
+		exp     string
+	}
+	tests := []testCase{
+		{false, "nosystemcrypto"},
+		{true, "systemcrypto"},
+	}
+	switch runtime.GOOS {
+	case "linux":
+		tests = append(tests, testCase{true, "opensslcrypto"})
+	case "windows":
+		switch runtime.GOARCH {
+		case "amd64", "arm64":
+			tests = append(tests, testCase{true, "cngcrypto"})
+		}
+	case "darwin":
+		tests = append(tests, testCase{true, "darwincrypto"})
+	}
+	for _, tt := range tests {
+		t.Run(tt.exp, func(t *testing.T) {
+			t.Parallel()
+			out := outPath(t)
+
+			// Build the binary with the specified GOEXPERIMENT.
+			succed, _ := execGoTool(t, true, []string{"GOEXPERIMENT=" + tt.exp}, "build", "-o", out, file)
+			if !succed {
+				t.Skip("skipping test because GOEXPERIMENT not supported by the current toolchain")
+			}
+
+			// Check that the binary has the correct settings.
+			_, settings := execGoTool(t, false, nil, "version", "-m", out)
+			hashSetting := strings.Contains(settings, "microsoft_systemcrypto=1")
+			if tt.enabled && !hashSetting {
+				t.Errorf("expected microsoft_systemcrypto=1 in settings, got %v", settings)
+			} else if !tt.enabled && hashSetting {
+				t.Errorf("expected microsoft_systemcrypto=1 not to be in settings, got %v", settings)
+			}
+		})
+	}
+}
diff --git a/src/cmd/go/testdata/script/env_cross_build.txt b/src/cmd/go/testdata/script/env_cross_build.txt
index 91d1cb936d3652..b7f503d9506324 100644
--- a/src/cmd/go/testdata/script/env_cross_build.txt
+++ b/src/cmd/go/testdata/script/env_cross_build.txt
@@ -1,6 +1,8 @@
 # Test that the correct default GOEXPERIMENT is used when cross
 # building with GOENV (#46815).
 
+[GOEXPERIMENT:systemcrypto] skip # systemcrypto is not supported on ios nor mips
+
 # Unset variables set by the TestScript harness. Users typically won't
 # explicitly configure these, and #46815 doesn't repro if they are.
 env GOOS=
diff --git a/src/cmd/go/testdata/script/test_android_issue62123.txt b/src/cmd/go/testdata/script/test_android_issue62123.txt
index 2f46a6b44bffe1..69071289e2de81 100644
--- a/src/cmd/go/testdata/script/test_android_issue62123.txt
+++ b/src/cmd/go/testdata/script/test_android_issue62123.txt
@@ -1,3 +1,5 @@
+[GOEXPERIMENT:systemcrypto] skip # systemcrypto is not supported on android
+
 env GOOS=android GOARCH=amd64 CGO_ENABLED=0
 
 ! go build -o $devnull cmd/buildid
diff --git a/src/cmd/internal/testdir/testdir_test.go b/src/cmd/internal/testdir/testdir_test.go
index 483a9ec33c6798..fad922e3f844ab 100644
--- a/src/cmd/internal/testdir/testdir_test.go
+++ b/src/cmd/internal/testdir/testdir_test.go
@@ -118,6 +118,13 @@ func Test(t *testing.T) {
 	goDebug = env.GODEBUG
 	tmpDir = t.TempDir()
 
+	if slices.ContainsFunc(strings.Split(goExperiment, ","), func(s string) bool {
+		return s == "systemcrypto" || s == "opensslcrypto" || s == "cngcrypto" || s == "darwincrypto"
+	}) {
+		// Codegen tests cross-compile to platforms that do not support system crypto.
+		*allCodegen = false
+	}
+
 	common := testCommon{
 		gorootTestDir: filepath.Join(testenv.GOROOT(t), "test"),
 		runoutputGate: make(chan bool, *runoutputLimit),
diff --git a/src/cmd/link/link_test.go b/src/cmd/link/link_test.go
index 2fe32600f1d6c7..f84cbdc5ee25f0 100644
--- a/src/cmd/link/link_test.go
+++ b/src/cmd/link/link_test.go
@@ -9,6 +9,7 @@ import (
 	"bytes"
 	"debug/macho"
 	"errors"
+	"internal/goexperiment"
 	"internal/platform"
 	"internal/testenv"
 	"os"
@@ -280,6 +281,9 @@ func TestBuildForTvOS(t *testing.T) {
 	if runtime.GOOS != "darwin" {
 		t.Skip("skipping on non-darwin platform")
 	}
+	if goexperiment.SystemCrypto {
+		t.Skip("tvOS is not supported by system crypto")
+	}
 	if testing.Short() && os.Getenv("GO_BUILDER_NAME") == "" {
 		t.Skip("skipping in -short mode with $GO_BUILDER_NAME empty")
 	}
@@ -512,6 +516,10 @@ func TestIssue34788Android386TLSSequence(t *testing.T) {
 		t.Skip("skipping on non-{linux,darwin}/amd64 platform")
 	}
 
+	if goexperiment.SystemCrypto {
+		t.Skip("skipping on system crypto, which does not support android/386")
+	}
+
 	t.Parallel()
 
 	tmpdir := t.TempDir()
diff --git a/src/crypto/internal/backend/backend_test.go b/src/crypto/internal/backend/backend_test.go
new file mode 100644
index 00000000000000..c2c06d3bff8c74
--- /dev/null
+++ b/src/crypto/internal/backend/backend_test.go
@@ -0,0 +1,30 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package backend
+
+import (
+	"testing"
+)
+
+// Test that Unreachable panics.
+func TestUnreachable(t *testing.T) {
+	defer func() {
+		if Enabled {
+			if err := recover(); err == nil {
+				t.Fatal("expected Unreachable to panic")
+			}
+		} else {
+			if err := recover(); err != nil {
+				t.Fatalf("expected Unreachable to be a no-op")
+			}
+		}
+	}()
+	Unreachable()
+}
+
+// Test that UnreachableExceptTests does not panic (this is a test).
+func TestUnreachableExceptTests(t *testing.T) {
+	UnreachableExceptTests()
+}
diff --git a/src/crypto/internal/backend/bbig/big.go b/src/crypto/internal/backend/bbig/big.go
new file mode 100644
index 00000000000000..ab3f30825dcfa1
--- /dev/null
+++ b/src/crypto/internal/backend/bbig/big.go
@@ -0,0 +1,17 @@
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !goexperiment.systemcrypto || (goexperiment.darwincrypto && !cgo)
+
+package bbig
+
+import "math/big"
+
+func Enc(b *big.Int) []uint {
+	return nil
+}
+
+func Dec(b []uint) *big.Int {
+	return nil
+}
diff --git a/src/crypto/internal/backend/bbig/big_cng.go b/src/crypto/internal/backend/bbig/big_cng.go
new file mode 100644
index 00000000000000..92623031fd87d0
--- /dev/null
+++ b/src/crypto/internal/backend/bbig/big_cng.go
@@ -0,0 +1,12 @@
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.cngcrypto
+
+package bbig
+
+import "github.com/microsoft/go-crypto-winnative/cng/bbig"
+
+var Enc = bbig.Enc
+var Dec = bbig.Dec
diff --git a/src/crypto/internal/backend/bbig/big_darwin.go b/src/crypto/internal/backend/bbig/big_darwin.go
new file mode 100644
index 00000000000000..77f3ca5d262769
--- /dev/null
+++ b/src/crypto/internal/backend/bbig/big_darwin.go
@@ -0,0 +1,12 @@
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.darwincrypto && cgo
+
+package bbig
+
+import "github.com/microsoft/go-crypto-darwin/bbig"
+
+var Enc = bbig.Enc
+var Dec = bbig.Dec
diff --git a/src/crypto/internal/backend/bbig/big_openssl.go b/src/crypto/internal/backend/bbig/big_openssl.go
new file mode 100644
index 00000000000000..e6695dd66b1d02
--- /dev/null
+++ b/src/crypto/internal/backend/bbig/big_openssl.go
@@ -0,0 +1,12 @@
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.opensslcrypto
+
+package bbig
+
+import "github.com/golang-fips/openssl/v2/bbig"
+
+var Enc = bbig.Enc
+var Dec = bbig.Dec
diff --git a/src/crypto/internal/backend/cng_windows.go b/src/crypto/internal/backend/cng_windows.go
new file mode 100644
index 00000000000000..c5bffa7b6c3949
--- /dev/null
+++ b/src/crypto/internal/backend/cng_windows.go
@@ -0,0 +1,345 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.cngcrypto && windows
+
+// Package cng provides access to CNGCrypto implementation functions.
+// Check the variable Enabled to find out whether CNGCrypto is available.
+// If CNGCrypto is not available, the functions in this package all panic.
+package backend
+
+import (
+	"crypto"
+	"crypto/cipher"
+	"crypto/internal/backend/fips140"
+	"crypto/internal/boring/sig"
+	"hash"
+	_ "unsafe"
+
+	"github.com/microsoft/go-crypto-winnative/cng"
+)
+
+func init() {
+	fn := func() bool {
+		enabled, err := cng.FIPS()
+		if err != nil {
+			panic("cngcrypto: unknown FIPS mode: " + err.Error())
+		}
+		return enabled
+	}
+	if err := fips140.Check(fn); err != nil {
+		panic("cngcrypto: " + err.Error())
+	}
+	sig.BoringCrypto()
+}
+
+// Enabled controls whether FIPS crypto is enabled.
+const Enabled = true
+
+type BigInt = cng.BigInt
+
+const RandReader = cng.RandReader
+
+func SupportsHash(h crypto.Hash) bool {
+	return cng.SupportsHash(h)
+}
+
+func SupportsCurve(curve string) bool            { return true }
+func SupportsRSAOAEPLabel(label []byte) bool     { return true }
+func SupportsPKCS1v15Hash(hash crypto.Hash) bool { return true }
+
+func NewMD5() hash.Hash        { return cng.NewMD5() }
+func NewSHA1() hash.Hash       { return cng.NewSHA1() }
+func NewSHA224() hash.Hash     { panic("cngcrypto: not available") }
+func NewSHA256() hash.Hash     { return cng.NewSHA256() }
+func NewSHA384() hash.Hash     { return cng.NewSHA384() }
+func NewSHA512() hash.Hash     { return cng.NewSHA512() }
+func NewSHA512_224() hash.Hash { panic("cngcrypto: not available") }
+func NewSHA512_256() hash.Hash { panic("cngcrypto: not available") }
+
+func MD5(p []byte) (sum [16]byte)        { return cng.MD5(p) }
+func SHA1(p []byte) (sum [20]byte)       { return cng.SHA1(p) }
+func SHA224(p []byte) (sum [28]byte)     { panic("cngcrypto: not available") }
+func SHA256(p []byte) (sum [32]byte)     { return cng.SHA256(p) }
+func SHA384(p []byte) (sum [48]byte)     { return cng.SHA384(p) }
+func SHA512(p []byte) (sum [64]byte)     { return cng.SHA512(p) }
+func SHA512_224(p []byte) (sum [28]byte) { panic("cngcrypto: not available") }
+func SHA512_256(p []byte) (sum [32]byte) { panic("cngcrypto: not available") }
+
+func NewHMAC(h func() hash.Hash, key []byte) hash.Hash {
+	return cng.NewHMAC(h, key)
+}
+
+func NewAESCipher(key []byte) (cipher.Block, error) {
+	return cng.NewAESCipher(key)
+}
+
+func NewGCMTLS(c cipher.Block) (cipher.AEAD, error) {
+	return cng.NewGCMTLS(c)
+}
+
+func NewGCMTLS13(c cipher.Block) (cipher.AEAD, error) {
+	return cng.NewGCMTLS13(c)
+}
+
+type PublicKeyECDSA = cng.PublicKeyECDSA
+type PrivateKeyECDSA = cng.PrivateKeyECDSA
+
+func GenerateKeyECDSA(curve string) (X, Y, D cng.BigInt, err error) {
+	return cng.GenerateKeyECDSA(curve)
+}
+
+func NewPrivateKeyECDSA(curve string, X, Y, D cng.BigInt) (*cng.PrivateKeyECDSA, error) {
+	return cng.NewPrivateKeyECDSA(curve, X, Y, D)
+}
+
+func NewPublicKeyECDSA(curve string, X, Y cng.BigInt) (*cng.PublicKeyECDSA, error) {
+	return cng.NewPublicKeyECDSA(curve, X, Y)
+}
+
+//go:linkname encodeSignature crypto/ecdsa.encodeSignature
+func encodeSignature(r, s []byte) ([]byte, error)
+
+//go:linkname parseSignature crypto/ecdsa.parseSignature
+func parseSignature(sig []byte) (r, s []byte, err error)
+
+func SignMarshalECDSA(priv *cng.PrivateKeyECDSA, hash []byte) ([]byte, error) {
+	r, s, err := cng.SignECDSA(priv, hash)
+	if err != nil {
+		return nil, err
+	}
+	return encodeSignature(r, s)
+}
+
+func VerifyECDSA(pub *cng.PublicKeyECDSA, hash []byte, sig []byte) bool {
+	rBytes, sBytes, err := parseSignature(sig)
+	if err != nil {
+		return false
+	}
+	return cng.VerifyECDSA(pub, hash, cng.BigInt(rBytes), cng.BigInt(sBytes))
+}
+
+func SignECDSA(priv *cng.PrivateKeyECDSA, hash []byte) (r, s cng.BigInt, err error) {
+	return cng.SignECDSA(priv, hash)
+}
+
+func VerifyECDSARaw(pub *cng.PublicKeyECDSA, hash []byte, r, s cng.BigInt) bool {
+	return cng.VerifyECDSA(pub, hash, r, s)
+}
+
+func SupportsRSAKeyPrimes(primes int) bool {
+	// CNG only supports 2-prime RSA keys.
+	return primes == 2
+}
+
+func SupportsRSASaltLength(sign bool, salt int) bool {
+	if sign {
+		return true
+	}
+	return salt != 0 // rsa.PSSSaltLengthAuto
+}
+
+type PublicKeyRSA = cng.PublicKeyRSA
+type PrivateKeyRSA = cng.PrivateKeyRSA
+
+func DecryptRSAOAEP(h, mgfHash hash.Hash, priv *cng.PrivateKeyRSA, ciphertext, label []byte) ([]byte, error) {
+	return cng.DecryptRSAOAEP(h, priv, ciphertext, label)
+}
+
+func DecryptRSAPKCS1(priv *cng.PrivateKeyRSA, ciphertext []byte) ([]byte, error) {
+	return cng.DecryptRSAPKCS1(priv, ciphertext)
+}
+
+func DecryptRSANoPadding(priv *cng.PrivateKeyRSA, ciphertext []byte) ([]byte, error) {
+	return cng.DecryptRSANoPadding(priv, ciphertext)
+}
+
+func EncryptRSAOAEP(h, mgfHash hash.Hash, pub *cng.PublicKeyRSA, msg, label []byte) ([]byte, error) {
+	return cng.EncryptRSAOAEP(h, pub, msg, label)
+}
+
+func EncryptRSAPKCS1(pub *cng.PublicKeyRSA, msg []byte) ([]byte, error) {
+	return cng.EncryptRSAPKCS1(pub, msg)
+}
+
+func EncryptRSANoPadding(pub *cng.PublicKeyRSA, msg []byte) ([]byte, error) {
+	return cng.EncryptRSANoPadding(pub, msg)
+}
+
+func GenerateKeyRSA(bits int) (N, E, D, P, Q, Dp, Dq, Qinv cng.BigInt, err error) {
+	return cng.GenerateKeyRSA(bits)
+}
+
+func NewPrivateKeyRSA(N, E, D, P, Q, Dp, Dq, Qinv cng.BigInt) (*cng.PrivateKeyRSA, error) {
+	return cng.NewPrivateKeyRSA(N, E, D, P, Q, Dp, Dq, Qinv)
+}
+
+func NewPublicKeyRSA(N, E cng.BigInt) (*cng.PublicKeyRSA, error) {
+	return cng.NewPublicKeyRSA(N, E)
+}
+
+func SignRSAPKCS1v15(priv *cng.PrivateKeyRSA, h crypto.Hash, hashed []byte) ([]byte, error) {
+	return cng.SignRSAPKCS1v15(priv, h, hashed)
+}
+
+func SignRSAPSS(priv *cng.PrivateKeyRSA, h crypto.Hash, hashed []byte, saltLen int) ([]byte, error) {
+	return cng.SignRSAPSS(priv, h, hashed, saltLen)
+}
+
+func VerifyRSAPKCS1v15(pub *cng.PublicKeyRSA, h crypto.Hash, hashed, sig []byte) error {
+	return cng.VerifyRSAPKCS1v15(pub, h, hashed, sig)
+}
+
+func VerifyRSAPSS(pub *cng.PublicKeyRSA, h crypto.Hash, hashed, sig []byte, saltLen int) error {
+	return cng.VerifyRSAPSS(pub, h, hashed, sig, saltLen)
+}
+
+type PrivateKeyECDH = cng.PrivateKeyECDH
+type PublicKeyECDH = cng.PublicKeyECDH
+
+func ECDH(priv *cng.PrivateKeyECDH, pub *cng.PublicKeyECDH) ([]byte, error) {
+	return cng.ECDH(priv, pub)
+}
+
+func GenerateKeyECDH(curve string) (*cng.PrivateKeyECDH, []byte, error) {
+	return cng.GenerateKeyECDH(curve)
+}
+
+func NewPrivateKeyECDH(curve string, bytes []byte) (*cng.PrivateKeyECDH, error) {
+	return cng.NewPrivateKeyECDH(curve, bytes)
+}
+
+func NewPublicKeyECDH(curve string, bytes []byte) (*cng.PublicKeyECDH, error) {
+	return cng.NewPublicKeyECDH(curve, bytes)
+}
+
+func SupportsTLS13KDF() bool {
+	return false
+}
+
+func ExpandTLS13KDF(h func() hash.Hash, pseudorandomKey, label, context []byte, keyLength int) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func SupportsHKDF() bool {
+	return cng.SupportsHKDF()
+}
+
+func ExpandHKDF(h func() hash.Hash, pseudorandomKey, info []byte, keyLength int) ([]byte, error) {
+	return cng.ExpandHKDF(h, pseudorandomKey, info, keyLength)
+}
+
+func ExtractHKDF(h func() hash.Hash, secret, salt []byte) ([]byte, error) {
+	return cng.ExtractHKDF(h, secret, salt)
+}
+
+func SupportsPBKDF2() bool { return true }
+
+func PBKDF2(password, salt []byte, iter, keyLen int, h func() hash.Hash) ([]byte, error) {
+	return cng.PBKDF2(password, salt, iter, keyLen, h)
+}
+
+func SupportsTLS1PRF() bool {
+	return true
+}
+
+func TLS1PRF(result, secret, label, seed []byte, h func() hash.Hash) error {
+	return cng.TLS1PRF(result, secret, label, seed, h)
+}
+
+func SupportsDESCipher() bool {
+	return true
+}
+
+func SupportsTripleDESCipher() bool {
+	return true
+}
+
+func NewDESCipher(key []byte) (cipher.Block, error) {
+	return cng.NewDESCipher(key)
+}
+
+func NewTripleDESCipher(key []byte) (cipher.Block, error) {
+	return cng.NewTripleDESCipher(key)
+}
+
+func SupportsRC4() bool { return true }
+
+type RC4Cipher = cng.RC4Cipher
+
+func NewRC4Cipher(key []byte) (*RC4Cipher, error) { return cng.NewRC4Cipher(key) }
+
+func SupportsEd25519() bool { return false }
+
+type PublicKeyEd25519 struct{}
+
+func (k PublicKeyEd25519) Bytes() ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+type PrivateKeyEd25519 struct{}
+
+func (k PrivateKeyEd25519) Bytes() ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func GenerateKeyEd25519() (PrivateKeyEd25519, error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPrivateKeyEd25519(priv []byte) (PrivateKeyEd25519, error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPublicKeyEd25519(pub []byte) (PublicKeyEd25519, error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPrivateKeyEd25519FromSeed(seed []byte) (PrivateKeyEd25519, error) {
+	panic("cryptobackend: not available")
+}
+
+func SignEd25519(priv PrivateKeyEd25519, message []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func VerifyEd25519(pub PublicKeyEd25519, message, sig []byte) error {
+	panic("cryptobackend: not available")
+}
+
+type PrivateKeyDSA = cng.PrivateKeyDSA
+type PublicKeyDSA = cng.PublicKeyDSA
+
+func SupportsDSA(l, n int) bool {
+	// These are the only N values supported by CNG
+	return n == 160 || n == 256
+}
+
+func GenerateParametersDSA(l, n int) (p, q, g cng.BigInt, err error) {
+	params, err := cng.GenerateParametersDSA(l)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return params.P, params.Q, params.G, nil
+}
+
+func GenerateKeyDSA(p, q, g cng.BigInt) (x, y cng.BigInt, err error) {
+	return cng.GenerateKeyDSA(cng.DSAParameters{P: p, Q: q, G: g})
+}
+
+func NewPrivateKeyDSA(p, q, g, x, y cng.BigInt) (*cng.PrivateKeyDSA, error) {
+	return cng.NewPrivateKeyDSA(cng.DSAParameters{P: p, Q: q, G: g}, x, y)
+}
+
+func NewPublicKeyDSA(p, q, g, y cng.BigInt) (*cng.PublicKeyDSA, error) {
+	return cng.NewPublicKeyDSA(cng.DSAParameters{P: p, Q: q, G: g}, y)
+}
+
+func SignDSA(priv *PrivateKeyDSA, hash []byte, parseSignature func([]byte) (cng.BigInt, cng.BigInt, error)) (r, s cng.BigInt, err error) {
+	return cng.SignDSA(priv, hash)
+}
+
+func VerifyDSA(pub *PublicKeyDSA, hashed []byte, r, s cng.BigInt, encodeSignature func(r, s cng.BigInt) ([]byte, error)) bool {
+	return cng.VerifyDSA(pub, hashed, r, s)
+}
diff --git a/src/crypto/internal/backend/common.go b/src/crypto/internal/backend/common.go
new file mode 100644
index 00000000000000..a9682181ffa154
--- /dev/null
+++ b/src/crypto/internal/backend/common.go
@@ -0,0 +1,46 @@
+// Copyright 2022 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package backend
+
+import (
+	"crypto/internal/boring/sig"
+	"runtime"
+)
+
+// Unreachable marks code that should be unreachable
+// when backend is in use.
+func Unreachable() {
+	if Enabled {
+		panic("cryptobackend: invalid code execution")
+	} else {
+		// Code that's unreachable is exactly the code
+		// we want to detect for reporting standard Go crypto.
+		sig.StandardCrypto()
+	}
+}
+
+// Provided by runtime.crypto_backend_runtime_arg0 to avoid os import.
+func runtime_arg0() string
+
+func hasSuffix(s, t string) bool {
+	return len(s) > len(t) && s[len(s)-len(t):] == t
+}
+
+// UnreachableExceptTests marks code that should be unreachable
+// when backend is in use. It panics.
+func UnreachableExceptTests() {
+	// runtime_arg0 is not supported on windows.
+	// We are going through the same code patch on linux,
+	// so if we are unintentionally calling an 'unreachable' function,
+	// we will catch it there.
+	if Enabled && runtime.GOOS != "windows" {
+		name := runtime_arg0()
+		// If ran on Windows we'd need to allow _test.exe and .test.exe as well.
+		if !hasSuffix(name, "_test") && !hasSuffix(name, ".test") {
+			println("cryptobackend: unexpected code execution in", name)
+			panic("cryptobackend: invalid code execution")
+		}
+	}
+}
diff --git a/src/crypto/internal/backend/darwin_darwin.go b/src/crypto/internal/backend/darwin_darwin.go
new file mode 100644
index 00000000000000..fea0f284de2439
--- /dev/null
+++ b/src/crypto/internal/backend/darwin_darwin.go
@@ -0,0 +1,372 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.darwincrypto && darwin && cgo
+
+// Package darwin provides access to DarwinCrypto implementation functions.
+// Check the variable Enabled to find out whether DarwinCrypto is available.
+// If DarwinCrypto is not available, the functions in this package all panic.
+package backend
+
+import (
+	"crypto"
+	"crypto/cipher"
+	"crypto/internal/backend/fips140"
+	"crypto/internal/boring/sig"
+	"crypto/internal/fips140/nistec"
+	"errors"
+	"hash"
+	_ "unsafe"
+
+	"github.com/microsoft/go-crypto-darwin/xcrypto"
+)
+
+func init() {
+	// Darwin is considered FIPS compliant.
+	if err := fips140.Check(func() bool { return true }); err != nil {
+		panic("darwincrypto: " + err.Error())
+	}
+	sig.BoringCrypto()
+}
+
+// Enabled controls whether FIPS crypto is enabled.
+const Enabled = true
+
+type BigInt = xcrypto.BigInt
+
+const RandReader = xcrypto.RandReader
+
+func SupportsHash(h crypto.Hash) bool {
+	return xcrypto.SupportsHash(h)
+}
+
+func SupportsCurve(curve string) bool {
+	switch curve {
+	case "P-256", "P-384", "P-521":
+		return true
+	}
+	return false
+}
+
+func SupportsRSAOAEPLabel(label []byte) bool {
+	// CommonCrypto doesn't support labels
+	// https://github.com/microsoft/go-crypto-darwin/issues/22
+	return len(label) == 0
+}
+
+func SupportsPKCS1v15Hash(hash crypto.Hash) bool {
+	switch hash {
+	case crypto.SHA1, crypto.SHA224, crypto.SHA256, crypto.SHA384, crypto.SHA512, 0:
+		return true
+	}
+	return false
+}
+
+func NewMD5() hash.Hash    { return xcrypto.NewMD5() }
+func NewSHA1() hash.Hash   { return xcrypto.NewSHA1() }
+func NewSHA224() hash.Hash { panic("cryptobackend: not available") }
+func NewSHA256() hash.Hash { return xcrypto.NewSHA256() }
+func NewSHA384() hash.Hash { return xcrypto.NewSHA384() }
+func NewSHA512() hash.Hash { return xcrypto.NewSHA512() }
+
+func MD5(p []byte) (sum [16]byte)        { return xcrypto.MD5(p) }
+func SHA1(p []byte) (sum [20]byte)       { return xcrypto.SHA1(p) }
+func SHA224(p []byte) (sum [28]byte)     { panic("cryptobackend: not available") }
+func SHA256(p []byte) (sum [32]byte)     { return xcrypto.SHA256(p) }
+func SHA384(p []byte) (sum [48]byte)     { return xcrypto.SHA384(p) }
+func SHA512(p []byte) (sum [64]byte)     { return xcrypto.SHA512(p) }
+func SHA512_224(p []byte) (sum [28]byte) { panic("cryptobackend: not available") }
+func SHA512_256(p []byte) (sum [32]byte) { panic("cryptobackend: not available") }
+
+func NewHMAC(h func() hash.Hash, key []byte) hash.Hash {
+	return xcrypto.NewHMAC(h, key)
+}
+
+func NewAESCipher(key []byte) (cipher.Block, error) {
+	return xcrypto.NewAESCipher(key)
+}
+
+func NewGCMTLS(c cipher.Block) (cipher.AEAD, error) {
+	return xcrypto.NewGCMTLS(c)
+}
+
+func NewGCMTLS13(c cipher.Block) (cipher.AEAD, error) {
+	return xcrypto.NewGCMTLS13(c)
+}
+
+type PublicKeyECDSA = xcrypto.PublicKeyECDSA
+type PrivateKeyECDSA = xcrypto.PrivateKeyECDSA
+
+func GenerateKeyECDSA(curve string) (X, Y, D xcrypto.BigInt, err error) {
+	return xcrypto.GenerateKeyECDSA(curve)
+}
+
+func NewPrivateKeyECDSA(curve string, X, Y, D xcrypto.BigInt) (*xcrypto.PrivateKeyECDSA, error) {
+	return xcrypto.NewPrivateKeyECDSA(curve, X, Y, D)
+}
+
+func NewPublicKeyECDSA(curve string, X, Y xcrypto.BigInt) (*xcrypto.PublicKeyECDSA, error) {
+	return xcrypto.NewPublicKeyECDSA(curve, X, Y)
+}
+
+//go:linkname encodeSignature crypto/ecdsa.encodeSignature
+func encodeSignature(r, s []byte) ([]byte, error)
+
+//go:linkname parseSignature crypto/ecdsa.parseSignature
+func parseSignature(sig []byte) (r, s []byte, err error)
+
+func SignMarshalECDSA(priv *xcrypto.PrivateKeyECDSA, hash []byte) ([]byte, error) {
+	return xcrypto.SignMarshalECDSA(priv, hash)
+}
+
+func VerifyECDSA(pub *xcrypto.PublicKeyECDSA, hash []byte, sig []byte) bool {
+	return xcrypto.VerifyECDSA(pub, hash, sig)
+}
+
+func SupportsRSAKeyPrimes(primes int) bool {
+	return primes == 2
+}
+
+func SupportsRSASaltLength(sign bool, salt int) bool {
+	// CommonCrypto doesn't support custom salt length
+	return salt == -1
+}
+
+type PublicKeyRSA = xcrypto.PublicKeyRSA
+type PrivateKeyRSA = xcrypto.PrivateKeyRSA
+
+func DecryptRSAOAEP(h, mgfHash hash.Hash, priv *xcrypto.PrivateKeyRSA, ciphertext, label []byte) ([]byte, error) {
+	return xcrypto.DecryptRSAOAEP(h, priv, ciphertext, label)
+}
+
+func DecryptRSAPKCS1(priv *xcrypto.PrivateKeyRSA, ciphertext []byte) ([]byte, error) {
+	return xcrypto.DecryptRSAPKCS1(priv, ciphertext)
+}
+
+func DecryptRSANoPadding(priv *xcrypto.PrivateKeyRSA, ciphertext []byte) ([]byte, error) {
+	return xcrypto.DecryptRSANoPadding(priv, ciphertext)
+}
+
+func EncryptRSAOAEP(h, mgfHash hash.Hash, pub *xcrypto.PublicKeyRSA, msg, label []byte) ([]byte, error) {
+	return xcrypto.EncryptRSAOAEP(h, pub, msg, label)
+}
+
+func EncryptRSAPKCS1(pub *xcrypto.PublicKeyRSA, msg []byte) ([]byte, error) {
+	return xcrypto.EncryptRSAPKCS1(pub, msg)
+}
+
+func EncryptRSANoPadding(pub *xcrypto.PublicKeyRSA, msg []byte) ([]byte, error) {
+	return xcrypto.EncryptRSANoPadding(pub, msg)
+}
+
+//go:linkname decodeKeyRSA crypto/rsa.decodeKey
+func decodeKeyRSA(data []byte) (N, E, D, P, Q, Dp, Dq, Qinv xcrypto.BigInt, err error)
+
+//go:linkname encodeKeyRSA crypto/rsa.encodeKey
+func encodeKeyRSA(N, E, D, P, Q, Dp, Dq, Qinv xcrypto.BigInt) ([]byte, error)
+
+//go:linkname encodePublicKeyRSA crypto/rsa.encodePublicKey
+func encodePublicKeyRSA(N, E xcrypto.BigInt) ([]byte, error)
+
+func GenerateKeyRSA(bits int) (N, E, D, P, Q, Dp, Dq, Qinv xcrypto.BigInt, err error) {
+	data, err := xcrypto.GenerateKeyRSA(bits)
+	if err != nil {
+		return
+	}
+	return decodeKeyRSA(data)
+}
+
+func NewPrivateKeyRSA(N, E, D, P, Q, Dp, Dq, Qinv xcrypto.BigInt) (*xcrypto.PrivateKeyRSA, error) {
+	encoded, err := encodeKeyRSA(N, E, D, P, Q, Dp, Dq, Qinv)
+	if err != nil {
+		return nil, err
+	}
+	return xcrypto.NewPrivateKeyRSA(encoded)
+}
+
+func NewPublicKeyRSA(N, E xcrypto.BigInt) (*xcrypto.PublicKeyRSA, error) {
+	encoded, err := encodePublicKeyRSA(N, E)
+	if err != nil {
+		return nil, err
+	}
+	return xcrypto.NewPublicKeyRSA(encoded)
+}
+
+func SignRSAPKCS1v15(priv *xcrypto.PrivateKeyRSA, h crypto.Hash, hashed []byte) ([]byte, error) {
+	return xcrypto.SignRSAPKCS1v15(priv, h, hashed)
+}
+
+func SignRSAPSS(priv *xcrypto.PrivateKeyRSA, h crypto.Hash, hashed []byte, saltLen int) ([]byte, error) {
+	return xcrypto.SignRSAPSS(priv, h, hashed, saltLen)
+}
+
+func VerifyRSAPKCS1v15(pub *xcrypto.PublicKeyRSA, h crypto.Hash, hashed, sig []byte) error {
+	return xcrypto.VerifyRSAPKCS1v15(pub, h, hashed, sig)
+}
+
+func VerifyRSAPSS(pub *xcrypto.PublicKeyRSA, h crypto.Hash, hashed, sig []byte, saltLen int) error {
+	return xcrypto.VerifyRSAPSS(pub, h, hashed, sig, saltLen)
+}
+
+type PrivateKeyECDH = xcrypto.PrivateKeyECDH
+type PublicKeyECDH = xcrypto.PublicKeyECDH
+
+func ECDH(priv *xcrypto.PrivateKeyECDH, pub *xcrypto.PublicKeyECDH) ([]byte, error) {
+	return xcrypto.ECDH(priv, pub)
+}
+
+func GenerateKeyECDH(curve string) (*xcrypto.PrivateKeyECDH, []byte, error) {
+	return xcrypto.GenerateKeyECDH(curve)
+}
+
+func NewPrivateKeyECDH(curve string, bytes []byte) (*xcrypto.PrivateKeyECDH, error) {
+	var key []byte
+	switch curve {
+	case "P-256":
+		p, err := nistec.NewP256Point().ScalarBaseMult(bytes)
+		if err != nil {
+			return nil, err
+		}
+		key = p.Bytes()
+	case "P-384":
+		p, err := nistec.NewP384Point().ScalarBaseMult(bytes)
+		if err != nil {
+			return nil, err
+		}
+		key = p.Bytes()
+	case "P-521":
+		p, err := nistec.NewP521Point().ScalarBaseMult(bytes)
+		if err != nil {
+			return nil, err
+		}
+		key = p.Bytes()
+	default:
+		return nil, errors.New("NewPrivateKeyECDH: unsupported curve: " + curve)
+	}
+	return xcrypto.NewPrivateKeyECDH(curve, key, bytes)
+}
+
+func NewPublicKeyECDH(curve string, bytes []byte) (*xcrypto.PublicKeyECDH, error) {
+	return xcrypto.NewPublicKeyECDH(curve, bytes)
+}
+
+func SupportsTLS13KDF() bool {
+	return false
+}
+
+func ExpandTLS13KDF(h func() hash.Hash, pseudorandomKey, label, context []byte, keyLength int) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func SupportsHKDF() bool {
+	return true
+}
+
+func ExpandHKDF(h func() hash.Hash, pseudorandomKey, info []byte, keyLength int) ([]byte, error) {
+	return xcrypto.ExpandHKDF(h, pseudorandomKey, info, keyLength)
+}
+
+func ExtractHKDF(h func() hash.Hash, secret, salt []byte) ([]byte, error) {
+	return xcrypto.ExtractHKDF(h, secret, salt)
+}
+
+func SupportsPBKDF2() bool {
+	return true
+}
+
+func PBKDF2(pass, salt []byte, iter, keyLen int, h func() hash.Hash) ([]byte, error) {
+	return xcrypto.PBKDF2(pass, salt, iter, keyLen, h)
+}
+
+func SupportsTLS1PRF() bool {
+	return false
+}
+
+func TLS1PRF(result, secret, label, seed []byte, h func() hash.Hash) error {
+	panic("cryptobackend: not available")
+}
+
+func SupportsDESCipher() bool {
+	return true
+}
+
+func SupportsTripleDESCipher() bool {
+	return true
+}
+
+func NewDESCipher(key []byte) (cipher.Block, error) {
+	return xcrypto.NewDESCipher(key)
+}
+
+func NewTripleDESCipher(key []byte) (cipher.Block, error) {
+	return xcrypto.NewTripleDESCipher(key)
+}
+
+func SupportsRC4() bool { return true }
+
+type RC4Cipher = xcrypto.RC4Cipher
+
+func NewRC4Cipher(key []byte) (*RC4Cipher, error) { return xcrypto.NewRC4Cipher(key) }
+
+func SupportsEd25519() bool {
+	return true
+}
+
+type PublicKeyEd25519 = xcrypto.PublicKeyEd25519
+type PrivateKeyEd25519 = xcrypto.PrivateKeyEd25519
+
+func GenerateKeyEd25519() (PrivateKeyEd25519, error) {
+	return xcrypto.GenerateKeyEd25519(), nil
+}
+
+func NewPrivateKeyEd25519(priv []byte) (PrivateKeyEd25519, error) {
+	return xcrypto.NewPrivateKeyEd25519(priv)
+}
+
+func NewPublicKeyEd25519(pub []byte) (PublicKeyEd25519, error) {
+	return xcrypto.NewPublicKeyEd25519(pub)
+}
+
+func NewPrivateKeyEd25519FromSeed(seed []byte) (PrivateKeyEd25519, error) {
+	return xcrypto.NewPrivateKeyEd25519FromSeed(seed)
+}
+
+func SignEd25519(priv PrivateKeyEd25519, message []byte) ([]byte, error) {
+	return xcrypto.SignEd25519(priv, message)
+}
+
+func VerifyEd25519(pub PublicKeyEd25519, message, sig []byte) error {
+	return xcrypto.VerifyEd25519(pub, message, sig)
+}
+
+func SupportsDSA(l, n int) bool {
+	return false
+}
+
+func GenerateParametersDSA(l, n int) (p, q, g xcrypto.BigInt, err error) {
+	panic("cryptobackend: not available")
+}
+
+type PrivateKeyDSA struct{}
+type PublicKeyDSA struct{}
+
+func GenerateKeyDSA(p, q, g xcrypto.BigInt) (x, y xcrypto.BigInt, err error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPrivateKeyDSA(p, q, g, x, y xcrypto.BigInt) (*PrivateKeyDSA, error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPublicKeyDSA(p, q, g, y xcrypto.BigInt) (*PublicKeyDSA, error) {
+	panic("cryptobackend: not available")
+}
+
+func SignDSA(priv *PrivateKeyDSA, hash []byte, parseSignature func([]byte) (xcrypto.BigInt, xcrypto.BigInt, error)) (r, s xcrypto.BigInt, err error) {
+	panic("cryptobackend: not available")
+}
+
+func VerifyDSA(pub *PublicKeyDSA, hashed []byte, r, s xcrypto.BigInt, encodeSignature func(r, s xcrypto.BigInt) ([]byte, error)) bool {
+	panic("cryptobackend: not available")
+}
diff --git a/src/crypto/internal/backend/fips140/cng.go b/src/crypto/internal/backend/fips140/cng.go
new file mode 100644
index 00000000000000..e4a1ddeeb7dc09
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/cng.go
@@ -0,0 +1,35 @@
+// Copyright 2024 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.cngcrypto
+
+package fips140
+
+import (
+	"internal/syscall/windows/sysdll"
+	"syscall"
+	"unsafe"
+)
+
+const backendEnabled = true
+
+// Don't use github.com/microsoft/go-crypto-winnative here.
+// The fips140 package should have minimal dependencies.
+// Also, don't directly query the system FIPS mode from the registry,
+// there are some no-longer documented legacy entries that can enable FIPS mode,
+// and BCryptGetFipsAlgorithmMode supports them all.
+var (
+	bcrypt = syscall.MustLoadDLL(sysdll.Add("bcrypt.dll"))
+
+	bcryptGetFipsAlgorithmMode = bcrypt.MustFindProc("BCryptGetFipsAlgorithmMode")
+)
+
+func systemFIPSMode() bool {
+	var enabled uint32
+	ret, _, _ := bcryptGetFipsAlgorithmMode.Call(uintptr(unsafe.Pointer(&enabled)))
+	if ret != 0 {
+		return false
+	}
+	return enabled != 0
+}
diff --git a/src/crypto/internal/backend/fips140/darwin.go b/src/crypto/internal/backend/fips140/darwin.go
new file mode 100644
index 00000000000000..8a2efe2346e84c
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/darwin.go
@@ -0,0 +1,13 @@
+// Copyright 2024 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.darwincrypto
+
+package fips140
+
+const backendEnabled = true
+
+func systemFIPSMode() bool {
+	return false
+}
diff --git a/src/crypto/internal/backend/fips140/fips140.go b/src/crypto/internal/backend/fips140/fips140.go
new file mode 100644
index 00000000000000..1bf7c3b7d63da8
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/fips140.go
@@ -0,0 +1,70 @@
+// Copyright 2024 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package fips140
+
+import (
+	"errors"
+	"internal/godebug"
+	"runtime"
+	"syscall"
+)
+
+var fips140GODEBUG = godebug.New("fips140")
+
+// Enabled reports whether FIPS 140 mode is enabled by using GODEBUG, GOFIPS, GOLANG_FIPS,
+// the 'requirefips' build tag, or any other platform-specific mechanism.
+func Enabled() bool {
+	return enabled
+}
+
+var enabled bool
+
+// message is a human-readable message about how [Enabled] was set.
+var message string
+
+func init() {
+	// TODO: Decide which environment variable to use.
+	// See https://github.com/microsoft/go/issues/397.
+	var ok bool
+	value := fips140GODEBUG.Value()
+	if value == "on" || value == "only" || value == "debug" {
+		message = "environment variable GODEBUG=fips140=" + value
+		value = "1"
+	} else if value, ok = syscall.Getenv("GOFIPS"); ok {
+		message = "environment variable GOFIPS"
+	} else if value, ok = syscall.Getenv("GOLANG_FIPS"); ok {
+		message = "environment variable GOLANG_FIPS"
+	} else if systemFIPSMode() {
+		message = "system FIPS mode"
+		value = "1"
+	}
+	enabled = value == "1"
+	if isRequireFIPS {
+		if isSkipFIPSCheck {
+			panic("the 'requirefips' build tag is enabled, but it conflicts " +
+				"with the 'ms_skipfipscheck' build tag")
+		}
+		message = "requirefips tag set"
+		enabled = true
+	}
+}
+
+// Check checks whether fips mode, as returned by fips(),
+// is valid given the current settings.
+func Check(fips func() bool) error {
+	if isSkipFIPSCheck || !enabled {
+		return nil
+	}
+	if !backendEnabled {
+		if runtime.GOOS != "linux" && runtime.GOOS != "windows" && runtime.GOOS != "darwin" {
+			return errors.New("FIPS mode requested (" + message + ") but no crypto backend is supported on " + runtime.GOOS)
+		}
+		return errors.New("FIPS mode requested (" + message + ") but no supported crypto backend is enabled")
+	}
+	if !fips() {
+		return errors.New("FIPS mode requested (" + message + ") but not available")
+	}
+	return nil
+}
diff --git a/src/crypto/internal/backend/fips140/isrequirefips.go b/src/crypto/internal/backend/fips140/isrequirefips.go
new file mode 100644
index 00000000000000..b33d08c84e2dae
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/isrequirefips.go
@@ -0,0 +1,9 @@
+// Copyright 2024 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build requirefips
+
+package fips140
+
+const isRequireFIPS = true
\ No newline at end of file
diff --git a/src/crypto/internal/backend/fips140/norequirefips.go b/src/crypto/internal/backend/fips140/norequirefips.go
new file mode 100644
index 00000000000000..6f01b9a3524dee
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/norequirefips.go
@@ -0,0 +1,9 @@
+// Copyright 2024 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !requirefips
+
+package fips140
+
+const isRequireFIPS = false
\ No newline at end of file
diff --git a/src/crypto/internal/backend/fips140/nosystemcrypto.go b/src/crypto/internal/backend/fips140/nosystemcrypto.go
new file mode 100644
index 00000000000000..289d6594eac54b
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/nosystemcrypto.go
@@ -0,0 +1,13 @@
+// Copyright 2024 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !goexperiment.systemcrypto
+
+package fips140
+
+const backendEnabled = false
+
+func systemFIPSMode() bool {
+	return false
+}
diff --git a/src/crypto/internal/backend/fips140/openssl_cgo.go b/src/crypto/internal/backend/fips140/openssl_cgo.go
new file mode 100644
index 00000000000000..a24db78572f528
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/openssl_cgo.go
@@ -0,0 +1,59 @@
+// Copyright 2024 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.opensslcrypto && cgo
+
+package fips140
+
+import (
+	_ "crypto/internal/backend/internal/opensslsetup"
+	"syscall"
+
+	"github.com/golang-fips/openssl/v2"
+)
+
+const backendEnabled = true
+
+// systemFIPSMode reports whether the system is in FIPS mode.
+// It first checks the kernel, and if that is not available, it checks the
+// OpenSSL library.
+func systemFIPSMode() bool {
+	if kernelFIPSMode() {
+		return true
+	}
+	return openssl.FIPS()
+}
+
+// kernelFIPSMode reports whether the kernel is in FIPS mode.
+func kernelFIPSMode() bool {
+	var fd int
+	for {
+		var err error
+		fd, err = syscall.Open("/proc/sys/crypto/fips_enabled", syscall.O_RDONLY, 0)
+		if err == nil {
+			break
+		}
+		switch err {
+		case syscall.EINTR:
+			continue
+		case syscall.ENOENT:
+			return false
+		default:
+			// If there is an error reading we could either panic or assume FIPS is not enabled.
+			// Panicking would be too disruptive for apps that don't require FIPS.
+			// If an app wants to be 100% sure that is running in FIPS mode
+			// it should use fips140.Enabled() or GODEBUG=fips140=1.
+			return false
+		}
+	}
+	defer syscall.Close(fd)
+	var tmp [1]byte
+	n, err := syscall.Read(fd, tmp[:])
+	if n != 1 || err != nil {
+		// We return false instead of panicing for the same reason as before.
+		return false
+	}
+	// fips_enabled can be either '0' or '1'.
+	return tmp[0] == '1'
+}
diff --git a/src/crypto/internal/backend/fips140/openssl_nocgo.go b/src/crypto/internal/backend/fips140/openssl_nocgo.go
new file mode 100644
index 00000000000000..a8ff666b6d182e
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/openssl_nocgo.go
@@ -0,0 +1,17 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.opensslcrypto && !cgo
+
+// The opensslcrypto experiment can be used even when cgo
+// is disabled. The OpenSSL backend can't be imported
+// without cgo, so we use a stub systemFIPSMode implementation.
+
+package fips140
+
+const backendEnabled = true
+
+func systemFIPSMode() bool {
+	return false
+}
diff --git a/src/crypto/internal/backend/fips140/requirefips_nosystemcrypto.go b/src/crypto/internal/backend/fips140/requirefips_nosystemcrypto.go
new file mode 100644
index 00000000000000..dad6bcea9451e2
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/requirefips_nosystemcrypto.go
@@ -0,0 +1,15 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build requirefips && !goexperiment.systemcrypto
+
+package fips140
+
+func init() {
+	`
+	The requirefips tag is enabled, but no crypto backend is enabled.
+	A crypto backend is required to enable FIPS mode.
+	For more information, visit https://github.com/microsoft/go/tree/microsoft/main/eng/doc/fips
+	`
+}
diff --git a/src/crypto/internal/backend/fips140/skipfipscheck_off.go b/src/crypto/internal/backend/fips140/skipfipscheck_off.go
new file mode 100644
index 00000000000000..f60b10dfa3aaf3
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/skipfipscheck_off.go
@@ -0,0 +1,9 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !ms_skipfipscheck
+
+package fips140
+
+const isSkipFIPSCheck = false
diff --git a/src/crypto/internal/backend/fips140/skipfipscheck_on.go b/src/crypto/internal/backend/fips140/skipfipscheck_on.go
new file mode 100644
index 00000000000000..d3d39fd77f98db
--- /dev/null
+++ b/src/crypto/internal/backend/fips140/skipfipscheck_on.go
@@ -0,0 +1,9 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build ms_skipfipscheck
+
+package fips140
+
+const isSkipFIPSCheck = true
diff --git a/src/crypto/internal/backend/internal/opensslsetup/opensslsetup.go b/src/crypto/internal/backend/internal/opensslsetup/opensslsetup.go
new file mode 100644
index 00000000000000..c6c1f53e7452ab
--- /dev/null
+++ b/src/crypto/internal/backend/internal/opensslsetup/opensslsetup.go
@@ -0,0 +1,70 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.opensslcrypto && cgo
+
+// opensslsetup is a package that initializes the OpenSSL library.
+// It doesn't export any symbol, but blank importing it has the
+// side effect of initializing the OpenSSL library.
+package opensslsetup
+
+import (
+	"syscall"
+
+	"github.com/golang-fips/openssl/v2"
+)
+
+// knownVersions is a list of supported and well-known libcrypto.so suffixes in decreasing version order.
+// FreeBSD library version numbering does not directly align to the version of OpenSSL.
+// Its preferred search order is 11 -> 111.
+// Some distributions use 1.0.0 and others (such as Debian) 1.0.2 to refer to the same OpenSSL 1.0.2 version.
+// Fedora derived distros use different naming for the version 1.0.x.
+var knownVersions = [...]string{"3", "1.1", "11", "111", "1.0.2", "1.0.0", "10"}
+
+const lcryptoPrefix = "libcrypto.so."
+
+func init() {
+	lib := library()
+	if err := openssl.Init(lib); err != nil {
+		panic("opensslcrypto: can't initialize OpenSSL " + lib + ": " + err.Error())
+	}
+}
+
+// library returns the name of the OpenSSL library to use.
+// It first checks the environment variable GO_OPENSSL_VERSION_OVERRIDE.
+// If that is not set, it searches a well-known list of library names.
+// If no library is found, it returns "libcrypto.so".
+func library() string {
+	if version, _ := syscall.Getenv("GO_OPENSSL_VERSION_OVERRIDE"); version != "" {
+		return lcryptoPrefix + version
+	}
+	if lib := searchKnownLibrary(); lib != "" {
+		return lib
+	}
+	return lcryptoPrefix[:len(lcryptoPrefix)-1] // no version found, try without version suffix
+}
+
+// checkVersion is a variable that holds the openssl.CheckVersion function.
+// It is initialized in the init function to allow overriding in tests.
+var checkVersion = openssl.CheckVersion
+
+// searchKnownLibrary returns the name of the highest available FIPS-enabled version of OpenSSL
+// using the known library suffixes.
+// If no FIPS-enabled version is found, it returns the name of the highest available version.
+// If no version is found, it returns an empty string.
+func searchKnownLibrary() string {
+	var lcryptoFallback string
+	for _, v := range knownVersions {
+		lcryptoCandidate := lcryptoPrefix + v
+		if exists, fips := checkVersion(lcryptoCandidate); exists {
+			if fips {
+				return lcryptoCandidate
+			}
+			if lcryptoFallback == "" {
+				lcryptoFallback = lcryptoCandidate
+			}
+		}
+	}
+	return lcryptoFallback
+}
diff --git a/src/crypto/internal/backend/internal/opensslsetup/opensslsetup_test.go b/src/crypto/internal/backend/internal/opensslsetup/opensslsetup_test.go
new file mode 100644
index 00000000000000..e9f67449710d65
--- /dev/null
+++ b/src/crypto/internal/backend/internal/opensslsetup/opensslsetup_test.go
@@ -0,0 +1,92 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.opensslcrypto && cgo
+
+package opensslsetup
+
+import "testing"
+
+func mockCheckVersion(t *testing.T, fn func(string) (bool, bool)) {
+	original := checkVersion
+	t.Cleanup(func() {
+		checkVersion = original
+	})
+	checkVersion = fn
+}
+
+func assertLibrary(t *testing.T, expected string) {
+	if result := library(); result != expected {
+		t.Errorf("expected %s, got %s", expected, result)
+	}
+}
+
+func TestLibraryWithEnvOverride(t *testing.T) {
+	t.Setenv("GO_OPENSSL_VERSION_OVERRIDE", "1.1")
+	mockCheckVersion(t, func(s string) (bool, bool) { return false, false })
+	assertLibrary(t, "libcrypto.so.1.1")
+}
+
+func TestLibraryWithKnownVersion(t *testing.T) {
+	t.Setenv("GO_OPENSSL_VERSION_OVERRIDE", "")
+
+	const maxLib = "libcrypto.so.3"
+
+	t.Run("AllExistsNoneFIPS", func(t *testing.T) {
+		mockCheckVersion(t, func(s string) (bool, bool) {
+			return true, false
+		})
+		for _, v := range knownVersions {
+			t.Run(v, func(t *testing.T) {
+				assertLibrary(t, maxLib)
+			})
+		}
+	})
+
+	t.Run("OnlyOneExists", func(t *testing.T) {
+		for _, v := range knownVersions {
+			t.Run(v, func(t *testing.T) {
+				expected := "libcrypto.so." + v
+				mockCheckVersion(t, func(s string) (bool, bool) {
+					if s == expected {
+						return true, false
+					}
+					return false, false
+				})
+				assertLibrary(t, expected)
+			})
+		}
+	})
+
+	t.Run("AllExistsOnlyOneFIPS", func(t *testing.T) {
+		fipsLib := "libcrypto.so.1.1"
+		mockCheckVersion(t, func(s string) (bool, bool) {
+			return true, s == fipsLib
+		})
+		for _, v := range knownVersions {
+			t.Run(v, func(t *testing.T) {
+				assertLibrary(t, fipsLib)
+			})
+		}
+	})
+
+	t.Run("AllExistsAndAreFIPS", func(t *testing.T) {
+		mockCheckVersion(t, func(s string) (bool, bool) {
+			return true, true
+		})
+		for _, v := range knownVersions {
+			t.Run(v, func(t *testing.T) {
+				assertLibrary(t, maxLib)
+			})
+		}
+	})
+}
+
+func TestLibraryNoVersionFound(t *testing.T) {
+	t.Setenv("GO_OPENSSL_VERSION_OVERRIDE", "")
+	mockCheckVersion(t, func(string) (bool, bool) {
+		return false, false
+	})
+	assertLibrary(t, "libcrypto.so")
+}
diff --git a/src/crypto/internal/backend/internal/opensslsetup/stub.go b/src/crypto/internal/backend/internal/opensslsetup/stub.go
new file mode 100644
index 00000000000000..3b92bfc5fbabf6
--- /dev/null
+++ b/src/crypto/internal/backend/internal/opensslsetup/stub.go
@@ -0,0 +1,8 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Placeholder to allow the opensslsetup package to be imported
+// without cgo enabled or without goexperiment.opensslcrypto.
+
+package opensslsetup
diff --git a/src/crypto/internal/backend/nobackend.go b/src/crypto/internal/backend/nobackend.go
new file mode 100644
index 00000000000000..ce267825dc6ddc
--- /dev/null
+++ b/src/crypto/internal/backend/nobackend.go
@@ -0,0 +1,251 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !goexperiment.systemcrypto || ((goexperiment.opensslcrypto || goexperiment.darwincrypto) && !cgo)
+
+package backend
+
+import (
+	"crypto"
+	"crypto/cipher"
+	"crypto/internal/backend/fips140"
+	"hash"
+)
+
+func init() {
+	if err := fips140.Check(func() bool { return false }); err != nil {
+		panic(err)
+	}
+}
+
+const Enabled = false
+
+type BigInt = []uint
+
+type randReader int
+
+func (randReader) Read(b []byte) (int, error) { panic("cryptobackend: not available") }
+
+const RandReader = randReader(0)
+
+func SupportsHash(h crypto.Hash) bool { panic("cryptobackend: not available") }
+
+func SupportsCurve(curve string) bool            { panic("cryptobackend: not available") }
+func SupportsRSAOAEPLabel(label []byte) bool     { panic("cryptobackend: not available") }
+func SupportsPKCS1v15Hash(hash crypto.Hash) bool { panic("cryptobackend: not available") }
+
+func NewMD5() hash.Hash        { panic("cryptobackend: not available") }
+func NewSHA1() hash.Hash       { panic("cryptobackend: not available") }
+func NewSHA224() hash.Hash     { panic("cryptobackend: not available") }
+func NewSHA256() hash.Hash     { panic("cryptobackend: not available") }
+func NewSHA384() hash.Hash     { panic("cryptobackend: not available") }
+func NewSHA512() hash.Hash     { panic("cryptobackend: not available") }
+func NewSHA512_224() hash.Hash { panic("cryptobackend: not available") }
+func NewSHA512_256() hash.Hash { panic("cryptobackend: not available") }
+
+func MD5(p []byte) (sum [16]byte)        { panic("cryptobackend: not available") }
+func SHA1(p []byte) (sum [20]byte)       { panic("cryptobackend: not available") }
+func SHA224(p []byte) (sum [28]byte)     { panic("cryptobackend: not available") }
+func SHA256(p []byte) (sum [32]byte)     { panic("cryptobackend: not available") }
+func SHA384(p []byte) (sum [48]byte)     { panic("cryptobackend: not available") }
+func SHA512(p []byte) (sum [64]byte)     { panic("cryptobackend: not available") }
+func SHA512_224(p []byte) (sum [28]byte) { panic("cryptobackend: not available") }
+func SHA512_256(p []byte) (sum [32]byte) { panic("cryptobackend: not available") }
+
+func NewHMAC(h func() hash.Hash, key []byte) hash.Hash { panic("cryptobackend: not available") }
+
+func NewAESCipher(key []byte) (cipher.Block, error)   { panic("cryptobackend: not available") }
+func NewGCMTLS(c cipher.Block) (cipher.AEAD, error)   { panic("cryptobackend: not available") }
+func NewGCMTLS13(c cipher.Block) (cipher.AEAD, error) { panic("cryptobackend: not available") }
+
+type PublicKeyECDSA struct{ _ int }
+type PrivateKeyECDSA struct{ _ int }
+
+func GenerateKeyECDSA(curve string) (X, Y, D BigInt, err error) {
+	panic("cryptobackend: not available")
+}
+func NewPrivateKeyECDSA(curve string, X, Y, D BigInt) (*PrivateKeyECDSA, error) {
+	panic("cryptobackend: not available")
+}
+func NewPublicKeyECDSA(curve string, X, Y BigInt) (*PublicKeyECDSA, error) {
+	panic("cryptobackend: not available")
+}
+func SignMarshalECDSA(priv *PrivateKeyECDSA, hash []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func VerifyECDSA(pub *PublicKeyECDSA, hash []byte, sig []byte) bool {
+	panic("cryptobackend: not available")
+}
+
+func SupportsRSAKeyPrimes(primes int) bool           { panic("cryptobackend: not available") }
+func SupportsRSASaltLength(sign bool, salt int) bool { panic("cryptobackend: not available") }
+
+type PublicKeyRSA struct{ _ int }
+type PrivateKeyRSA struct{ _ int }
+
+func DecryptRSAOAEP(h, mgfHash hash.Hash, priv *PrivateKeyRSA, ciphertext, label []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func DecryptRSAPKCS1(priv *PrivateKeyRSA, ciphertext []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func DecryptRSANoPadding(priv *PrivateKeyRSA, ciphertext []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func EncryptRSAOAEP(h, mgfHash hash.Hash, pub *PublicKeyRSA, msg, label []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func EncryptRSAPKCS1(pub *PublicKeyRSA, msg []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func EncryptRSANoPadding(pub *PublicKeyRSA, msg []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func GenerateKeyRSA(bits int) (N, E, D, P, Q, Dp, Dq, Qinv BigInt, err error) {
+	panic("cryptobackend: not available")
+}
+func NewPrivateKeyRSA(N, E, D, P, Q, Dp, Dq, Qinv BigInt) (*PrivateKeyRSA, error) {
+	panic("cryptobackend: not available")
+}
+func NewPublicKeyRSA(N, E BigInt) (*PublicKeyRSA, error) {
+	panic("cryptobackend: not available")
+}
+func SignRSAPKCS1v15(priv *PrivateKeyRSA, h crypto.Hash, hashed []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func SignRSAPSS(priv *PrivateKeyRSA, h crypto.Hash, hashed []byte, saltLen int) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+func VerifyRSAPKCS1v15(pub *PublicKeyRSA, h crypto.Hash, hashed, sig []byte) error {
+	panic("cryptobackend: not available")
+}
+func VerifyRSAPSS(pub *PublicKeyRSA, h crypto.Hash, hashed, sig []byte, saltLen int) error {
+	panic("cryptobackend: not available")
+}
+
+type PublicKeyECDH struct{}
+type PrivateKeyECDH struct{}
+
+func ECDH(*PrivateKeyECDH, *PublicKeyECDH) ([]byte, error)    { panic("cryptobackend: not available") }
+func GenerateKeyECDH(string) (*PrivateKeyECDH, []byte, error) { panic("cryptobackend: not available") }
+func NewPrivateKeyECDH(string, []byte) (*PrivateKeyECDH, error) {
+	panic("cryptobackend: not available")
+}
+func NewPublicKeyECDH(string, []byte) (*PublicKeyECDH, error) { panic("cryptobackend: not available") }
+func (*PublicKeyECDH) Bytes() []byte                          { panic("cryptobackend: not available") }
+func (*PrivateKeyECDH) PublicKey() (*PublicKeyECDH, error)    { panic("cryptobackend: not available") }
+
+func SupportsTLS13KDF() bool { panic("cryptobackend: not available") }
+
+func ExpandTLS13KDF(h func() hash.Hash, pseudorandomKey, label, context []byte, keyLength int) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func SupportsHKDF() bool { panic("cryptobackend: not available") }
+
+func ExpandHKDF(h func() hash.Hash, pseudorandomKey, info []byte, keyLength int) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func ExtractHKDF(h func() hash.Hash, secret, salt []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func SupportsPBKDF2() bool { panic("cryptobackend: not available") }
+
+func PBKDF2(password, salt []byte, iter, keyLen int, fh func() hash.Hash) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func SupportsTLS1PRF() bool { panic("cryptobackend: not available") }
+
+func TLS1PRF(result, secret, label, seed []byte, h func() hash.Hash) error {
+	panic("cryptobackend: not available")
+}
+
+func SupportsDESCipher() bool { panic("cryptobackend: not available") }
+
+func SupportsTripleDESCipher() bool { panic("cryptobackend: not available") }
+
+func NewDESCipher(key []byte) (cipher.Block, error) { panic("cryptobackend: not available") }
+
+func NewTripleDESCipher(key []byte) (cipher.Block, error) { panic("cryptobackend: not available") }
+
+func SupportsRC4() bool { panic("cryptobackend: not available") }
+
+type RC4Cipher struct{}
+
+func (c *RC4Cipher) Reset()                       { panic("cryptobackend: not available") }
+func (c *RC4Cipher) XORKeyStream(dst, src []byte) { panic("cryptobackend: not available") }
+
+func NewRC4Cipher(key []byte) (*RC4Cipher, error) { panic("cryptobackend: not available") }
+
+func SupportsEd25519() bool { panic("cryptobackend: not available") }
+
+type PublicKeyEd25519 struct{}
+
+func (k PublicKeyEd25519) Bytes() ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+type PrivateKeyEd25519 struct{}
+
+func (k PrivateKeyEd25519) Bytes() ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func GenerateKeyEd25519() (PrivateKeyEd25519, error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPrivateKeyEd25519(priv []byte) (PrivateKeyEd25519, error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPublicKeyEd25519(pub []byte) (PublicKeyEd25519, error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPrivateKeyEd25519FromSeed(seed []byte) (PrivateKeyEd25519, error) {
+	panic("cryptobackend: not available")
+}
+
+func SignEd25519(priv PrivateKeyEd25519, message []byte) ([]byte, error) {
+	panic("cryptobackend: not available")
+}
+
+func VerifyEd25519(pub PublicKeyEd25519, message, sig []byte) error {
+	panic("cryptobackend: not available")
+}
+
+func SupportsDSA(l, n int) bool {
+	panic("cryptobackend: not available")
+}
+
+func GenerateParametersDSA(l, n int) (p, q, g BigInt, err error) {
+	panic("cryptobackend: not available")
+}
+
+type PublicKeyDSA struct{ _ int }
+type PrivateKeyDSA struct{ _ int }
+
+func GenerateKeyDSA(p, q, g BigInt) (x, y BigInt, err error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPrivateKeyDSA(p, q, g, x, y BigInt) (*PrivateKeyDSA, error) {
+	panic("cryptobackend: not available")
+}
+
+func NewPublicKeyDSA(p, q, g, y BigInt) (*PublicKeyDSA, error) {
+	panic("cryptobackend: not available")
+}
+
+func SignDSA(priv *PrivateKeyDSA, hash []byte, parseSignature func([]byte) (BigInt, BigInt, error)) (r, s BigInt, err error) {
+	panic("cryptobackend: not available")
+}
+
+func VerifyDSA(pub *PublicKeyDSA, hashed []byte, r, s BigInt, encodeSignature func(r, s BigInt) ([]byte, error)) bool {
+	panic("cryptobackend: not available")
+}
diff --git a/src/crypto/internal/backend/openssl_linux.go b/src/crypto/internal/backend/openssl_linux.go
new file mode 100644
index 00000000000000..02c3c67f822ae2
--- /dev/null
+++ b/src/crypto/internal/backend/openssl_linux.go
@@ -0,0 +1,332 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build goexperiment.opensslcrypto && linux && cgo
+
+// Package openssl provides access to OpenSSLCrypto implementation functions.
+// Check the variable Enabled to find out whether OpenSSLCrypto is available.
+// If OpenSSLCrypto is not available, the functions in this package all panic.
+package backend
+
+import (
+	"crypto"
+	"crypto/cipher"
+	"crypto/internal/backend/fips140"
+	_ "crypto/internal/backend/internal/opensslsetup"
+	"crypto/internal/boring/sig"
+	"hash"
+
+	"github.com/golang-fips/openssl/v2"
+)
+
+// Enabled controls whether FIPS crypto is enabled.
+const Enabled = true
+
+type BigInt = openssl.BigInt
+
+func init() {
+	// Some distributions, e.g. Azure Linux 3, don't set the `fips=yes` property when running in FIPS mode,
+	// but they configure OpenSSL to use a FIPS-compliant provider (in the case of Azure Linux 3, the SCOSSL provider).
+	// In this cases, openssl.FIPS would return `false` and openssl.FIPSCapable would return `true`.
+	// We don't care about the `fips=yes` property as long as the provider is FIPS-compliant, so use
+	// openssl.FIPSCapable to determine whether FIPS mode is enabled.
+	if err := fips140.Check(func() bool { return openssl.FIPSCapable() }); err != nil {
+		// This path can be reached for the following reasons:
+		// - In OpenSSL 1, the active engine doesn't support FIPS mode.
+		// - In OpenSSL 1, the active engine supports FIPS mode, but it is not enabled.
+		// - In OpenSSL 3, the provider used by default doesn't match the `fips=yes` query.
+		panic("opensslcrypto: " + err.Error() + ": " + openssl.VersionText())
+	}
+	sig.BoringCrypto()
+}
+
+const RandReader = openssl.RandReader
+
+func SupportsHash(h crypto.Hash) bool {
+	return openssl.SupportsHash(h)
+}
+
+func SupportsCurve(curve string) bool            { return true }
+func SupportsRSAOAEPLabel(label []byte) bool     { return true }
+func SupportsPKCS1v15Hash(hash crypto.Hash) bool { return true }
+
+func NewMD5() hash.Hash        { return openssl.NewMD5() }
+func NewSHA1() hash.Hash       { return openssl.NewSHA1() }
+func NewSHA224() hash.Hash     { return openssl.NewSHA224() }
+func NewSHA256() hash.Hash     { return openssl.NewSHA256() }
+func NewSHA384() hash.Hash     { return openssl.NewSHA384() }
+func NewSHA512() hash.Hash     { return openssl.NewSHA512() }
+func NewSHA512_224() hash.Hash { return openssl.NewSHA512_224() }
+func NewSHA512_256() hash.Hash { return openssl.NewSHA512_256() }
+
+func MD5(p []byte) (sum [16]byte)        { return openssl.MD5(p) }
+func SHA1(p []byte) (sum [20]byte)       { return openssl.SHA1(p) }
+func SHA224(p []byte) (sum [28]byte)     { return openssl.SHA224(p) }
+func SHA256(p []byte) (sum [32]byte)     { return openssl.SHA256(p) }
+func SHA384(p []byte) (sum [48]byte)     { return openssl.SHA384(p) }
+func SHA512(p []byte) (sum [64]byte)     { return openssl.SHA512(p) }
+func SHA512_224(p []byte) (sum [28]byte) { return openssl.SHA512_224(p) }
+func SHA512_256(p []byte) (sum [32]byte) { return openssl.SHA512_256(p) }
+
+func NewHMAC(h func() hash.Hash, key []byte) hash.Hash { return openssl.NewHMAC(h, key) }
+
+func NewAESCipher(key []byte) (cipher.Block, error)   { return openssl.NewAESCipher(key) }
+func NewGCMTLS(c cipher.Block) (cipher.AEAD, error)   { return openssl.NewGCMTLS(c) }
+func NewGCMTLS13(c cipher.Block) (cipher.AEAD, error) { return openssl.NewGCMTLS13(c) }
+
+type PublicKeyECDSA = openssl.PublicKeyECDSA
+type PrivateKeyECDSA = openssl.PrivateKeyECDSA
+
+func GenerateKeyECDSA(curve string) (X, Y, D openssl.BigInt, err error) {
+	return openssl.GenerateKeyECDSA(curve)
+}
+
+func NewPrivateKeyECDSA(curve string, X, Y, D openssl.BigInt) (*openssl.PrivateKeyECDSA, error) {
+	return openssl.NewPrivateKeyECDSA(curve, X, Y, D)
+}
+
+func NewPublicKeyECDSA(curve string, X, Y openssl.BigInt) (*openssl.PublicKeyECDSA, error) {
+	return openssl.NewPublicKeyECDSA(curve, X, Y)
+}
+
+func SignMarshalECDSA(priv *openssl.PrivateKeyECDSA, hash []byte) ([]byte, error) {
+	return openssl.SignMarshalECDSA(priv, hash)
+}
+
+func VerifyECDSA(pub *openssl.PublicKeyECDSA, hash []byte, sig []byte) bool {
+	return openssl.VerifyECDSA(pub, hash, sig)
+}
+
+func SupportsRSAKeyPrimes(primes int) bool {
+	// The built-in OpenSSL 3 providers and OpenSSL 1 do support n-prime RSA keys,
+	// but SCOSSL only supports 2-prime RSA keys.
+	// Only 2-prime RSA keys are FIPS compliant, other n having compatibility
+	// and security issues. Even crypto/rsa deprecated rsa.GenerateMultiPrimeKey as of Go 1.21.
+	// Given the above reasons, we only support what SCOSSL supports.
+	return primes == 2
+}
+
+func SupportsRSASaltLength(sign bool, salt int) bool {
+	return true
+}
+
+type PublicKeyRSA = openssl.PublicKeyRSA
+type PrivateKeyRSA = openssl.PrivateKeyRSA
+
+func DecryptRSAOAEP(h, mgfHash hash.Hash, priv *openssl.PrivateKeyRSA, ciphertext, label []byte) ([]byte, error) {
+	return openssl.DecryptRSAOAEP(h, mgfHash, priv, ciphertext, label)
+}
+
+func DecryptRSAPKCS1(priv *openssl.PrivateKeyRSA, ciphertext []byte) ([]byte, error) {
+	return openssl.DecryptRSAPKCS1(priv, ciphertext)
+}
+
+func DecryptRSANoPadding(priv *openssl.PrivateKeyRSA, ciphertext []byte) ([]byte, error) {
+	return openssl.DecryptRSANoPadding(priv, ciphertext)
+}
+
+func EncryptRSAOAEP(h, mgfHash hash.Hash, pub *openssl.PublicKeyRSA, msg, label []byte) ([]byte, error) {
+	return openssl.EncryptRSAOAEP(h, mgfHash, pub, msg, label)
+}
+
+func EncryptRSAPKCS1(pub *openssl.PublicKeyRSA, msg []byte) ([]byte, error) {
+	return openssl.EncryptRSAPKCS1(pub, msg)
+}
+
+func EncryptRSANoPadding(pub *openssl.PublicKeyRSA, msg []byte) ([]byte, error) {
+	return openssl.EncryptRSANoPadding(pub, msg)
+}
+
+func GenerateKeyRSA(bits int) (N, E, D, P, Q, Dp, Dq, Qinv openssl.BigInt, err error) {
+	return openssl.GenerateKeyRSA(bits)
+}
+
+func NewPrivateKeyRSA(N, E, D, P, Q, Dp, Dq, Qinv openssl.BigInt) (*openssl.PrivateKeyRSA, error) {
+	return openssl.NewPrivateKeyRSA(N, E, D, P, Q, Dp, Dq, Qinv)
+}
+
+func NewPublicKeyRSA(N, E openssl.BigInt) (*openssl.PublicKeyRSA, error) {
+	return openssl.NewPublicKeyRSA(N, E)
+}
+
+func SignRSAPKCS1v15(priv *openssl.PrivateKeyRSA, h crypto.Hash, hashed []byte) ([]byte, error) {
+	return openssl.SignRSAPKCS1v15(priv, h, hashed)
+}
+
+func SignRSAPSS(priv *openssl.PrivateKeyRSA, h crypto.Hash, hashed []byte, saltLen int) ([]byte, error) {
+	return openssl.SignRSAPSS(priv, h, hashed, saltLen)
+}
+
+func VerifyRSAPKCS1v15(pub *openssl.PublicKeyRSA, h crypto.Hash, hashed, sig []byte) error {
+	return openssl.VerifyRSAPKCS1v15(pub, h, hashed, sig)
+}
+
+func VerifyRSAPSS(pub *openssl.PublicKeyRSA, h crypto.Hash, hashed, sig []byte, saltLen int) error {
+	return openssl.VerifyRSAPSS(pub, h, hashed, sig, saltLen)
+}
+
+type PublicKeyECDH = openssl.PublicKeyECDH
+type PrivateKeyECDH = openssl.PrivateKeyECDH
+
+func ECDH(priv *openssl.PrivateKeyECDH, pub *openssl.PublicKeyECDH) ([]byte, error) {
+	return openssl.ECDH(priv, pub)
+}
+
+func GenerateKeyECDH(curve string) (*openssl.PrivateKeyECDH, []byte, error) {
+	return openssl.GenerateKeyECDH(curve)
+}
+
+func NewPrivateKeyECDH(curve string, bytes []byte) (*openssl.PrivateKeyECDH, error) {
+	return openssl.NewPrivateKeyECDH(curve, bytes)
+}
+
+func NewPublicKeyECDH(curve string, bytes []byte) (*openssl.PublicKeyECDH, error) {
+	return openssl.NewPublicKeyECDH(curve, bytes)
+}
+
+func SupportsTLS13KDF() bool {
+	return openssl.SupportsTLS13KDF()
+}
+
+func ExpandTLS13KDF(h func() hash.Hash, pseudorandomKey, label, context []byte, keyLength int) ([]byte, error) {
+	return openssl.ExpandTLS13KDF(h, pseudorandomKey, label, context, keyLength)
+}
+
+func SupportsHKDF() bool {
+	return openssl.SupportsHKDF()
+}
+
+func ExpandHKDF(h func() hash.Hash, pseudorandomKey, info []byte, keyLength int) ([]byte, error) {
+	return openssl.ExpandHKDFOneShot(h, pseudorandomKey, info, keyLength)
+}
+
+func ExtractHKDF(h func() hash.Hash, secret, salt []byte) ([]byte, error) {
+	return openssl.ExtractHKDF(h, secret, salt)
+}
+
+func SupportsPBKDF2() bool {
+	return openssl.SupportsPBKDF2()
+}
+
+func PBKDF2(pass, salt []byte, iter, keyLen int, h func() hash.Hash) ([]byte, error) {
+	return openssl.PBKDF2(pass, salt, iter, keyLen, h)
+}
+
+func SupportsTLS1PRF() bool {
+	return openssl.SupportsTLS1PRF()
+}
+
+func TLS1PRF(result, secret, label, seed []byte, h func() hash.Hash) error {
+	return openssl.TLS1PRF(result, secret, label, seed, h)
+}
+
+func SupportsDESCipher() bool {
+	return openssl.SupportsDESCipher()
+}
+
+func SupportsTripleDESCipher() bool {
+	return openssl.SupportsTripleDESCipher()
+}
+
+func NewDESCipher(key []byte) (cipher.Block, error) {
+	return openssl.NewDESCipher(key)
+}
+
+func NewTripleDESCipher(key []byte) (cipher.Block, error) {
+	return openssl.NewTripleDESCipher(key)
+}
+
+func SupportsRC4() bool {
+	return openssl.SupportsRC4()
+}
+
+type RC4Cipher = openssl.RC4Cipher
+
+func NewRC4Cipher(key []byte) (*RC4Cipher, error) { return openssl.NewRC4Cipher(key) }
+
+func SupportsEd25519() bool { return openssl.SupportsEd25519() }
+
+type PublicKeyEd25519 = *openssl.PublicKeyEd25519
+type PrivateKeyEd25519 = *openssl.PrivateKeyEd25519
+
+func GenerateKeyEd25519() (PrivateKeyEd25519, error) {
+	return openssl.GenerateKeyEd25519()
+}
+
+// Deprecated: use NewPrivateKeyEd25519 instead.
+func NewPrivateKeyEd25119(priv []byte) (PrivateKeyEd25519, error) {
+	return openssl.NewPrivateKeyEd25519(priv)
+}
+
+// Deprecated: use NewPublicKeyEd25519 instead.
+func NewPublicKeyEd25119(pub []byte) (PublicKeyEd25519, error) {
+	return openssl.NewPublicKeyEd25519(pub)
+}
+
+func NewPrivateKeyEd25519(priv []byte) (PrivateKeyEd25519, error) {
+	return openssl.NewPrivateKeyEd25519(priv)
+}
+
+func NewPublicKeyEd25519(pub []byte) (PublicKeyEd25519, error) {
+	return openssl.NewPublicKeyEd25519(pub)
+}
+
+func NewPrivateKeyEd25519FromSeed(seed []byte) (PrivateKeyEd25519, error) {
+	return openssl.NewPrivateKeyEd25519FromSeed(seed)
+}
+
+func SignEd25519(priv PrivateKeyEd25519, message []byte) ([]byte, error) {
+	return openssl.SignEd25519(priv, message)
+}
+
+func VerifyEd25519(pub PublicKeyEd25519, message, sig []byte) error {
+	return openssl.VerifyEd25519(pub, message, sig)
+}
+
+type PublicKeyDSA = openssl.PublicKeyDSA
+type PrivateKeyDSA = openssl.PrivateKeyDSA
+
+func SupportsDSA(l, n int) bool {
+	return openssl.SupportsDSA()
+}
+
+func GenerateParametersDSA(l, n int) (p, q, g openssl.BigInt, err error) {
+	params, err := openssl.GenerateParametersDSA(l, n)
+	return params.P, params.Q, params.G, err
+}
+
+func GenerateKeyDSA(p, q, g openssl.BigInt) (x, y openssl.BigInt, err error) {
+	return openssl.GenerateKeyDSA(openssl.DSAParameters{P: p, Q: q, G: g})
+}
+
+func NewPrivateKeyDSA(p, q, g, x, y openssl.BigInt) (*openssl.PrivateKeyDSA, error) {
+	return openssl.NewPrivateKeyDSA(openssl.DSAParameters{P: p, Q: q, G: g}, x, y)
+}
+
+func NewPublicKeyDSA(p, q, g, y openssl.BigInt) (*openssl.PublicKeyDSA, error) {
+	return openssl.NewPublicKeyDSA(openssl.DSAParameters{P: p, Q: q, G: g}, y)
+}
+
+func SignDSA(priv *PrivateKeyDSA, hash []byte, parseSignature func([]byte) (openssl.BigInt, openssl.BigInt, error)) (r, s openssl.BigInt, err error) {
+	sig, err := openssl.SignDSA(priv, hash)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	r, s, err = parseSignature(sig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return openssl.BigInt(r), openssl.BigInt(s), nil
+}
+
+func VerifyDSA(pub *PublicKeyDSA, hashed []byte, r, s openssl.BigInt, encodeSignature func(r, s openssl.BigInt) ([]byte, error)) bool {
+	sig, err := encodeSignature(r, s)
+	if err != nil {
+		return false
+	}
+
+	return openssl.VerifyDSA(pub, hashed, sig)
+}
diff --git a/src/crypto/internal/backend/stub.s b/src/crypto/internal/backend/stub.s
new file mode 100644
index 00000000000000..5e4b436554d44d
--- /dev/null
+++ b/src/crypto/internal/backend/stub.s
@@ -0,0 +1,10 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// runtime_arg0 is declared in common.go without a body.
+// It's provided by package runtime,
+// but the go command doesn't know that.
+// Having this assembly file keeps the go command
+// from complaining about the missing body
+// (because the implementation might be here).
diff --git a/src/crypto/systemcrypto_nocgo.go b/src/crypto/systemcrypto_nocgo.go
new file mode 100644
index 00000000000000..13782a8241b99a
--- /dev/null
+++ b/src/crypto/systemcrypto_nocgo.go
@@ -0,0 +1,15 @@
+// Copyright 2025 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build !cgo && (goexperiment.opensslcrypto || goexperiment.darwincrypto)
+
+package crypto
+
+func init() {
+	`
+	Using a crypto backend requires CGO_ENABLED=1.
+	
+	For more information, visit https://github.com/microsoft/go/tree/microsoft/main/eng/doc/fips
+	`
+}
diff --git a/src/go/build/deps_test.go b/src/go/build/deps_test.go
index 9b7613a62b5a31..1997d90f2f20ba 100644
--- a/src/go/build/deps_test.go
+++ b/src/go/build/deps_test.go
@@ -348,7 +348,7 @@ var depsRules = `
 	math/big, go/token
 	< go/constant;
 
-	FMT, internal/goexperiment
+	FMT, internal/goexperiment, internal/platform
 	< internal/buildcfg;
 
 	container/heap, go/constant, go/parser, internal/buildcfg, internal/goversion, internal/types/errors
@@ -537,6 +537,11 @@ var depsRules = `
 	< github.com/microsoft/go-crypto-winnative/internal/bcrypt
 	< github.com/microsoft/go-crypto-winnative/cng;
 
+	syscall,
+	github.com/golang-fips/openssl/v2
+	< crypto/internal/backend/internal/opensslsetup
+	< crypto/internal/backend/fips140;
+
 	crypto !< FIPS;
 
 	# CRYPTO is core crypto algorithms - no cgo, fmt, net.
@@ -545,16 +550,27 @@ var depsRules = `
 	NONE < crypto/internal/boring/sig, crypto/internal/boring/syso;
 	sync/atomic < crypto/internal/boring/bcache;
 
+	crypto,	crypto/cipher,
+	crypto/internal/boring/sig, crypto/internal/boring/syso
+	< crypto/internal/boring;
+
+	crypto/internal/boring/sig,
+	crypto/internal/backend/fips140,
+	github.com/golang-fips/openssl/v2,
+	github.com/microsoft/go-crypto-darwin/xcrypto,
+	github.com/microsoft/go-crypto-winnative/cng
+	< crypto/internal/backend;
+
 	FIPS, internal/godebug, embed,
 	crypto/internal/boring/sig,
 	crypto/internal/boring/syso,
 	crypto/internal/boring/bcache,
 	crypto/internal/fips140only,
 	crypto,
-	crypto/cipher
+	crypto/cipher,
+	crypto/internal/backend
 	< crypto/sha3
 	< crypto/internal/fips140hash
-	< crypto/internal/boring
 	< crypto/boring
 	< crypto/aes,
 	  crypto/des,
@@ -578,8 +594,14 @@ var depsRules = `
 	math/big, github.com/microsoft/go-crypto-darwin/xcrypto < github.com/microsoft/go-crypto-darwin/bbig;
 	math/big, github.com/microsoft/go-crypto-winnative/cng < github.com/microsoft/go-crypto-winnative/cng/bbig;
 
-	CRYPTO, FMT, math/big
-	< crypto/internal/boring/bbig
+	FMT, math/big, crypto/internal/boring
+	< crypto/internal/boring/bbig;
+
+	CRYPTO, FMT, math/big,
+	github.com/golang-fips/openssl/v2/bbig,
+	github.com/microsoft/go-crypto-darwin/bbig,
+	github.com/microsoft/go-crypto-winnative/cng/bbig
+	< crypto/internal/backend/bbig
 	< crypto/internal/fips140cache
 	< crypto/rand
 	< crypto/ed25519 # depends on crypto/rand.Reader
diff --git a/src/internal/buildcfg/exp.go b/src/internal/buildcfg/exp.go
index bba73cfab9a108..c6bc4f165e6048 100644
--- a/src/internal/buildcfg/exp.go
+++ b/src/internal/buildcfg/exp.go
@@ -5,11 +5,14 @@
 package buildcfg
 
 import (
+	"errors"
 	"fmt"
 	"reflect"
+	"slices"
 	"strings"
 
 	"internal/goexperiment"
+	goplatform "internal/platform"
 )
 
 // ExperimentFlags represents a set of GOEXPERIMENT flags relative to a baseline
@@ -167,6 +170,57 @@ func ParseGOEXPERIMENT(goos, goarch, goexp string) (*ExperimentFlags, error) {
 	if flags.BoringCrypto {
 		return nil, fmt.Errorf("GOEXPERIMENT boringcrypto is not supported in the Microsoft build of Go")
 	}
+	// Check crypto backend conflicts.
+	type platform struct {
+		goos   string
+		goarch []string // if nil or empty, enabled on all architectures
+	}
+	var backends = []struct {
+		name     string
+		enabled  bool
+		platform []platform
+	}{
+		{"systemcrypto", flags.SystemCrypto, []platform{
+			{"linux", nil},
+			{"windows", []string{"amd64", "arm64"}},
+			{"darwin", nil},
+		}},
+		{"opensslcrypto", flags.OpenSSLCrypto, []platform{{"linux", nil}}},
+		{"cngcrypto", flags.CNGCrypto, []platform{{"windows", []string{"amd64", "arm64"}}}},
+		{"darwincrypto", flags.DarwinCrypto, []platform{{"darwin", nil}}},
+	}
+	var errs []string
+	for i, backend := range backends {
+		if !backend.enabled {
+			continue
+		}
+		// Don't report an error if the build mode is not supported,
+		// since that is not a problem with the experiment configuration.
+		if goplatform.BuildModeSupported("gc", "default", goos, goarch) {
+			if !slices.ContainsFunc(backend.platform, func(p platform) bool { return p.goos == goos }) {
+				errs = append(errs, fmt.Sprintf("GOEXPERIMENT %s is not supported on %s", backend.name, goos))
+			} else {
+				for _, platform := range backend.platform {
+					if platform.goos == goos && (len(platform.goarch) > 0 && !slices.Contains(platform.goarch, goarch)) {
+						errs = append(errs, fmt.Sprintf("GOEXPERIMENT %s is not supported on %s/%s", backend.name, goos, goarch))
+					}
+				}
+			}
+		}
+		if backend.name != "systemcrypto" { // systemcrypto can coexist with other valid backends
+			for _, backend2 := range backends[i+1:] {
+				if !backend2.enabled {
+					continue
+				}
+				errs = append(errs, fmt.Sprintf("GOEXPERIMENT %s and %s are both enabled, but they are mutually exclusive",
+					backend.name, backend2.name))
+			}
+		}
+	}
+	if len(errs) > 0 {
+		errs = append(errs, "For more information, visit https://github.com/microsoft/go/tree/microsoft/main/eng/doc/fips")
+		return nil, errors.New(strings.Join(errs, "\n\t"))
+	}
 	return flags, nil
 }
 
diff --git a/src/runtime/runtime_boring.go b/src/runtime/runtime_boring.go
index 831ee67afc952d..d2f00d57c10286 100644
--- a/src/runtime/runtime_boring.go
+++ b/src/runtime/runtime_boring.go
@@ -14,3 +14,8 @@ func boring_runtime_arg0() string {
 	}
 	return argslice[0]
 }
+
+//go:linkname crypto_backend_runtime_arg0 crypto/internal/backend.runtime_arg0
+func crypto_backend_runtime_arg0() string {
+	return boring_runtime_arg0()
+}
diff --git a/src/syscall/exec_linux_test.go b/src/syscall/exec_linux_test.go
index 69d49169446d1c..e6d344ff67a8aa 100644
--- a/src/syscall/exec_linux_test.go
+++ b/src/syscall/exec_linux_test.go
@@ -300,7 +300,7 @@ func TestUnshareMountNameSpaceChroot(t *testing.T) {
 	})
 
 	cmd := testenv.Command(t, testenv.GoToolPath(t), "test", "-c", "-o", x, "syscall")
-	cmd.Env = append(cmd.Environ(), "CGO_ENABLED=0")
+	cmd.Env = append(cmd.Environ(), "CGO_ENABLED=0", "GOEXPERIMENT=nosystemcrypto")
 	if o, err := cmd.CombinedOutput(); err != nil {
 		t.Fatalf("%v: %v\n%s", cmd, err, o)
 	}
